<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Orchestrating Microbiome Analysis with Bioconductor (ECCB) • miaDemo</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Orchestrating Microbiome Analysis with Bioconductor (ECCB)">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">miaDemo</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/eccb_workflow.html">Orchestrating Microbiome Analysis with Bioconductor (ECCB)</a>
    </li>
    <li>
      <a href="../articles/mia_demo.html">Orchestrating Microbiome Analysis with Bioconductor (EuroBioC2024)</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/microbiome/miaDemo" class="external-link">
    <span class="fa fa-github"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Orchestrating Microbiome Analysis with
Bioconductor (ECCB)</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/microbiome/miaDemo/blob/devel/vignettes/eccb_workflow.Rmd" class="external-link"><code>vignettes/eccb_workflow.Rmd</code></a></small>
      <div class="hidden name"><code>eccb_workflow.Rmd</code></div>

    </div>

    
    
<p>Authors: Tuomas Borman, Pande Erawijantari, Giulio Benedetti,
Stefanie Peschel, Leo Lahti. <br> Last modified: 12 September,
2024.</p>
<p><img src="figures/mia_logo.png" width="150"></p>
<p>TODO:</p>
<ul>
<li>rename the workflow UPDATE DESCRIPTION</li>
</ul>
<p>THINGS TO CONSIDER - We should probably use same dataset. Data
integration part requires multiomics data. - I added idea for story but
feel free to modify.</p>
<p>PRELIMINARY SCHEDULE</p>
<p>13:00-13:20 Leo - Idea: Introduction and TreeSE - Introduction to
Bioconductor - Overview of data containers - Introduction to TreeSE
13:20-13:25 short Q&amp;A - Participants have time to explore on their
own 13:25-13:30 short break - Time to set up for the next session</p>
<p>13:30-13:50 Pande - Idea: After introducing TreeSE, we can now use
methods from mia. - What are diversity and composition? - How to
calculate and plot alpha diversity indices? - How to calculate and plot
PCoA(?) 13:50-13:55 short Q&amp;A - Participants have time to explore on
their own 13:55-14:00 short break - Time to set up for the next
session</p>
<p>14:00-14:20 Tuomas - Idea: Now participants know that we can utilize
methods from mia. Introduce slots and show how to handle complex data. -
reducedDim slot was briefly introduced in Pande’s session. (That
highlights the idea of slots hopefully since the result is the same as
already calculated.) - Introduce agglomerateByRanks() and altExp slot -
Add another omics to altExp and demonstrating column subsetting to show
sample linkage - Calculate cross-association? 14:20-14:30 short Q&amp;A
- Participants have time to explore on their own</p>
<p>15:00-15:20 Stefanie - Idea: Show what additional capabilities there
are - Maybe: we can calculate simple correlation network based on
cross-association results (Tuomas’s session) but this can be further
extended? 15:20-15:25 short Q&amp;A - Participants have time to explore
on their own 15:25-15:30 short break - Time to set up for the next
session</p>
<p>15:30-15:50 Giulio - Idea: Show what additional capabilities there
are - Maybe: the analyses performed earlier can be done in graphical
interface? 15:50-15:55 short Q&amp;A - Participants have time to explore
on their own 15:55-16:00 short break - Time to set up for the next
session</p>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<div class="section level3">
<h3 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h3>
<p>Learn how <em>miaverse</em> framework integrates microbiome data
science with the <em>SummarizedExperiment</em> ecosystem, and discover
the benefits this brings. We will cover key methods, explore <a href="https://microbiome.github.io/OMA/docs/devel" class="external-link">the OMA online
book</a>, and discuss the integration’s advantages.</p>
<p>We will start with a brief overview of the framework, followed by an
instructor-led live demonstration. The demonstration will showcase
selected key features and functionalities of <em>miaverse</em>. After
the demo, participant will have time to discuss the framework, ask
questions, and explore the tools and methods covered.</p>
<p>By the end of the workshop, attendees will have a understanding of
how to apply the framework to their microbiome research and will be
well-prepared to further explore the framework independently.</p>
</div>
<div class="section level3">
<h3 id="pre-requisites">Pre-requisites<a class="anchor" aria-label="anchor" href="#pre-requisites"></a>
</h3>
<ul>
<li>Basic knowledge of R coding</li>
<li>Familiarity with Bioconductor</li>
<li>Understanding of microbiome research</li>
</ul>
<p>If your time allows, we recommend to spend some time to explore
beforehand <a href="https://microbiome.github.io/OMA/docs/devel/" class="external-link">Orchestrating
Microbiome Analysis (OMA) online book</a>.</p>
</div>
<div class="section level3">
<h3 id="participation">Participation<a class="anchor" aria-label="anchor" href="#participation"></a>
</h3>
<p>Participants are encouraged to ask questions throughout the workshop.
The session will follow <a href="https://microbiome.github.io/OMA/docs/devel/pages/introductory_workflow.html" class="external-link">a
tutorial from the OMA online book</a>, with participants running the
tutorial alongside the instructor.</p>
<p>To facilitate this, Bioconductor provides <a href="https://workshop.bioconductor.org/" class="external-link">pre-installed virtual machines
with all necessary packages</a>.</p>
</div>
<div class="section level3">
<h3 id="r-bioconductor-packages-used">
<em>R</em> / <em>Bioconductor</em> packages used<a class="anchor" aria-label="anchor" href="#r-bioconductor-packages-used"></a>
</h3>
<p>In this workshop, we will focus on <a href="https://bioconductor.org/packages/devel/bioc/html/mia.html" class="external-link">the
<em>mia</em> package</a>, which is designed for microbiome data science.
The <em>mia</em> package provides essential tools for analyzing
microbiome data using the <em>SummarizedExperiment</em> framework.</p>
</div>
<div class="section level3">
<h3 id="time-outline">Time outline<a class="anchor" aria-label="anchor" href="#time-outline"></a>
</h3>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Background</td>
<td>10m</td>
</tr>
<tr class="even">
<td>Demo</td>
<td>20m</td>
</tr>
<tr class="odd">
<td>Questions and free exploration</td>
<td>10m</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="workshop-goals-and-objectives">Workshop goals and objectives<a class="anchor" aria-label="anchor" href="#workshop-goals-and-objectives"></a>
</h3>
<div class="section level4">
<h4 id="learning-goals">Learning goals<a class="anchor" aria-label="anchor" href="#learning-goals"></a>
</h4>
<ul>
<li><p><strong>Methods for microbiome analysis</strong>: Learn to
harness <em>SummarizedExperiment</em> ecosystem in your
project.</p></li>
<li><p><strong>Data structure</strong>: Understand how to utilize the
<em>TreeSummarizedExperiment</em> class for effective microbiome data
analysis.</p></li>
<li><p><strong>Access resources</strong>: Get familiar with additional
tools and resources, including <a href="https://microbiome.github.io/OMA/docs/devel" class="external-link">the OMA online
book</a>.</p></li>
</ul>
</div>
<div class="section level4">
<h4 id="learning-objectives">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h4>
<ul>
<li><p><strong>Analyze and apply methods</strong>: Apply the framework
to process and analyze microbiome data.</p></li>
<li><p><strong>Create visualizations</strong>: Generate and interpret
visualizations.</p></li>
<li><p><strong>Explore documentation</strong>: Use the <a href="https://microbiome.github.io/OMA/docs/devel" class="external-link">OMA</a> to explore
additional tools and methods.</p></li>
</ul>
</div>
</div>
</div>
<div class="section level2">
<h2 id="workshop">Workshop<a class="anchor" aria-label="anchor" href="#workshop"></a>
</h2>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In this workflow, we demonstrate how to analyze data from publicly
available microbiome resources using Bioconductor methods. We will cover
essential data wrangling steps, explore alpha and beta diversity
metrics, and introduce the fundamentals of data integration and
microbial network analysis. Additionally, we will explore how to utilize
the tools through a graphical interface.</p>
</div>
<div class="section level2">
<h2 id="bioconductor-resources-for-microbiome-data-science-leo">Bioconductor resources for microbiome data science (Leo)<a class="anchor" aria-label="anchor" href="#bioconductor-resources-for-microbiome-data-science-leo"></a>
</h2>
<p>Idea for the story:</p>
<p>” TreeSE is the primary data container used in miaverse framework.It
is extension to SE. It works like this. mia package has methods for
common operations. For instance, we can tranform the data. ”</p>
<div class="section level3">
<h3 id="load-packages">Load packages<a class="anchor" aria-label="anchor" href="#load-packages"></a>
</h3>
<p>First, we load the required packages into the session. The following
script ensures that any packages not already installed are automatically
installed.</p>
<pre><code><span><span class="co">#&gt; Loading required package: BiocManager</span></span>
<span><span class="co">#&gt; Loading required package: ComplexHeatmap</span></span>
<span><span class="co">#&gt; Loading required package: grid</span></span>
<span><span class="co">#&gt; ========================================</span></span>
<span><span class="co">#&gt; ComplexHeatmap version 2.21.0</span></span>
<span><span class="co">#&gt; Bioconductor page: http://bioconductor.org/packages/ComplexHeatmap/</span></span>
<span><span class="co">#&gt; Github page: https://github.com/jokergoo/ComplexHeatmap</span></span>
<span><span class="co">#&gt; Documentation: http://jokergoo.github.io/ComplexHeatmap-reference</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; If you use it in published research, please cite either one:</span></span>
<span><span class="co">#&gt; - Gu, Z. Complex Heatmap Visualization. iMeta 2022.</span></span>
<span><span class="co">#&gt; - Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional </span></span>
<span><span class="co">#&gt;     genomic data. Bioinformatics 2016.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; The new InteractiveComplexHeatmap package can directly export static </span></span>
<span><span class="co">#&gt; complex heatmaps into an interactive Shiny app with zero effort. Have a try!</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; This message can be suppressed by:</span></span>
<span><span class="co">#&gt;   suppressPackageStartupMessages(library(ComplexHeatmap))</span></span>
<span><span class="co">#&gt; ========================================</span></span>
<span><span class="co">#&gt; Loading required package: curatedMetagenomicData</span></span>
<span><span class="co">#&gt; Loading required package: SummarizedExperiment</span></span>
<span><span class="co">#&gt; Loading required package: MatrixGenerics</span></span>
<span><span class="co">#&gt; Loading required package: matrixStats</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'MatrixGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span></span>
<span><span class="co">#&gt;     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span></span>
<span><span class="co">#&gt;     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span></span>
<span><span class="co">#&gt;     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span></span>
<span><span class="co">#&gt;     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span></span>
<span><span class="co">#&gt;     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span></span>
<span><span class="co">#&gt;     colWeightedMeans, colWeightedMedians, colWeightedSds,</span></span>
<span><span class="co">#&gt;     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span></span>
<span><span class="co">#&gt;     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span></span>
<span><span class="co">#&gt;     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span></span>
<span><span class="co">#&gt;     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span></span>
<span><span class="co">#&gt;     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span></span>
<span><span class="co">#&gt;     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span></span>
<span><span class="co">#&gt;     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span></span>
<span><span class="co">#&gt;     rowWeightedSds, rowWeightedVars</span></span>
<span><span class="co">#&gt; Loading required package: GenomicRanges</span></span>
<span><span class="co">#&gt; Loading required package: stats4</span></span>
<span><span class="co">#&gt; Loading required package: BiocGenerics</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'BiocGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">#&gt;     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">#&gt;     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">#&gt;     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">#&gt;     Position, rank, rbind, Reduce, rownames, sapply, setdiff, table,</span></span>
<span><span class="co">#&gt;     tapply, union, unique, unsplit, which.max, which.min</span></span>
<span><span class="co">#&gt; Loading required package: S4Vectors</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'S4Vectors'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:utils':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     findMatches</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     expand.grid, I, unname</span></span>
<span><span class="co">#&gt; Loading required package: IRanges</span></span>
<span><span class="co">#&gt; Loading required package: GenomeInfoDb</span></span>
<span><span class="co">#&gt; Loading required package: Biobase</span></span>
<span><span class="co">#&gt; Welcome to Bioconductor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     Vignettes contain introductory material; view with</span></span>
<span><span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'Biobase'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:MatrixGenerics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     rowMedians</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyMissing, rowMedians</span></span>
<span><span class="co">#&gt; Loading required package: TreeSummarizedExperiment</span></span>
<span><span class="co">#&gt; Loading required package: SingleCellExperiment</span></span>
<span><span class="co">#&gt; Loading required package: Biostrings</span></span>
<span><span class="co">#&gt; Loading required package: XVector</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'Biostrings'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:grid':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     pattern</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     strsplit</span></span>
<span><span class="co">#&gt; Loading required package: mia</span></span>
<span><span class="co">#&gt; Loading required package: MultiAssayExperiment</span></span>
<span><span class="co">#&gt; This is mia version 1.13.38</span></span>
<span><span class="co">#&gt; - Online documentation and vignettes: https://microbiome.github.io/mia/</span></span>
<span><span class="co">#&gt; - Online book 'Orchestrating Microbiome Analysis (OMA)': https://microbiome.github.io/OMA/docs/devel/</span></span>
<span><span class="co">#&gt; Loading required package: miaViz</span></span>
<span><span class="co">#&gt; Loading required package: ggplot2</span></span>
<span><span class="co">#&gt; Loading required package: ggraph</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'miaViz'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:mia':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     plotNMDS</span></span>
<span><span class="co">#&gt; Loading required package: scater</span></span>
<span><span class="co">#&gt; Loading required package: scuttle</span></span>
<span><span class="co">#&gt; Loading required package: shadowtext</span></span></code></pre>
<p>If you encountered an error, install packages manually.</p>
</div>
<div class="section level3">
<h3 id="importing-data">Importing data<a class="anchor" aria-label="anchor" href="#importing-data"></a>
</h3>
<p>There are several openly available datasets listed in <span class="citation">[@sec-example-data]</span>. In this workflow, we will
retrieve data from the curatedMetagenomicData resource. We’ll focus on a
dataset that includes both colorectal cancer (CRC) patients and control
subjects. Our goal is to investigate whether CRC is associated with
alterations in the gut microbiota.</p>
<p>Let’s fetch the data from the database.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Study from which we want to fetch data</span></span>
<span><span class="va">study</span> <span class="op">&lt;-</span> <span class="st">"GuptaA_2019"</span></span>
<span><span class="co"># Experiments that we want to fetch</span></span>
<span><span class="va">experiments</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pathway_abundance"</span>, <span class="st">"relative_abundance"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span>collapse <span class="op">=</span> <span class="st">"|"</span><span class="op">)</span></span>
<span><span class="va">search_pattern</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">study</span>, <span class="st">".("</span>, <span class="va">experiments</span>, <span class="st">")"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Fetch data</span></span>
<span><span class="va">data_list</span> <span class="op">&lt;-</span> <span class="fu">curatedMetagenomicData</span><span class="op">(</span></span>
<span>    <span class="va">search_pattern</span>, dryrun <span class="op">=</span> <span class="cn">FALSE</span>, counts <span class="op">=</span> <span class="cn">TRUE</span>, rownames <span class="op">=</span> <span class="st">"short"</span><span class="op">)</span></span>
<span><span class="co">#&gt; $`2021-03-31.GuptaA_2019.relative_abundance`</span></span>
<span><span class="co">#&gt; dropping rows without rowTree matches:</span></span>
<span><span class="co">#&gt;   k__Bacteria|p__Actinobacteria|c__Coriobacteriia|o__Coriobacteriales|f__Coriobacteriaceae|g__Collinsella|s__Collinsella_stercoris</span></span>
<span><span class="co">#&gt;   k__Bacteria|p__Actinobacteria|c__Coriobacteriia|o__Coriobacteriales|f__Coriobacteriaceae|g__Enorma|s__[Collinsella]_massiliensis</span></span>
<span><span class="co">#&gt;   k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Carnobacteriaceae|g__Granulicatella|s__Granulicatella_elegans</span></span>
<span><span class="co">#&gt;   k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Ruminococcaceae|g__Ruminococcus|s__Ruminococcus_champanellensis</span></span>
<span><span class="co">#&gt;   k__Bacteria|p__Proteobacteria|c__Betaproteobacteria|o__Burkholderiales|f__Sutterellaceae|g__Sutterella|s__Sutterella_parvirubra</span></span>
<span></span>
<span><span class="co"># Add new names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">data_list</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pathway"</span>, <span class="st">"taxonomy"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># The assay name for taxonomy profile is incorrect since we have counts, not</span></span>
<span><span class="co"># relative abundances</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"taxonomy"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu">assayNames</span><span class="op">(</span><span class="va">tse</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"counts"</span></span>
<span><span class="va">data_list</span><span class="op">[[</span><span class="st">"taxonomy"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">tse</span></span>
<span></span>
<span><span class="co"># For pathway data, extract pathway information from rownames and add them to</span></span>
<span><span class="co"># rowData</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"pathway"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu">rowData</span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">[[</span><span class="st">"full_name"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span>
<span><span class="fu">rowData</span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">[[</span><span class="st">"pathway"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"\\|.*"</span>, <span class="st">""</span>, <span class="fu">rowData</span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">[[</span><span class="st">"full_name"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">data_list</span><span class="op">[[</span><span class="st">"pathway"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">tse</span></span>
<span></span>
<span><span class="co"># Get taxonomy object</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"taxonomy"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">tse</span></span>
<span><span class="co">#&gt; class: TreeSummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 304 60 </span></span>
<span><span class="co">#&gt; metadata(1): agglomerated_by_rank</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(304): [Bacteroides] pectinophilus [Clostridium] hylemonae ...</span></span>
<span><span class="co">#&gt;   Victivallis vadensis Weissella confusa</span></span>
<span><span class="co">#&gt; rowData names(7): superkingdom phylum ... genus species</span></span>
<span><span class="co">#&gt; colnames(60): GupDM_A_11 GupDM_A_15 ... GupDM_JO GupDM_JP</span></span>
<span><span class="co">#&gt; colData names(27): study_name subject_id ... disease_stage</span></span>
<span><span class="co">#&gt;   disease_location</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; rowLinks: a LinkDataFrame (304 rows)</span></span>
<span><span class="co">#&gt; rowTree: 1 phylo tree(s) (10430 leaves)</span></span>
<span><span class="co">#&gt; colLinks: NULL</span></span>
<span><span class="co">#&gt; colTree: NULL</span></span></code></pre></div>
<p>The output is a <code>TreeSummarizedExperiment</code>
(<code>TreeSE</code>) object. For details on how to handle and access
the data, refer to <span class="citation">[@sec-containers]</span>.</p>
</div>
<div class="section level3">
<h3 id="preprocess">Preprocess<a class="anchor" aria-label="anchor" href="#preprocess"></a>
</h3>
<p>A common step in microbiome workflows is agglomeration, where we
summarize data at specific taxonomy ranks. Additionally, preprocessing
involves transforming the data to address the unique characteristics of
microbiome profiling data, for instance. Details on agglomeration and
transformation are covered in <span class="citation">[@sec-agglomeration]</span> and <span class="citation">[@sec-assay-transform]</span>, respectively.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Agglomerate data</span></span>
<span><span class="va">tse_family</span> <span class="op">&lt;-</span> <span class="fu">agglomerateByRank</span><span class="op">(</span><span class="va">tse</span>, rank <span class="op">=</span> <span class="st">"family"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The following values are already present in `metadata` and will be</span></span>
<span><span class="co">#&gt; overwritten: 'agglomerated_by_rank'. Consider using the 'name' argument to</span></span>
<span><span class="co">#&gt; specify alternative names.</span></span>
<span></span>
<span><span class="co"># Transform the main TreeSE</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu">transformAssay</span><span class="op">(</span><span class="va">tse</span>, method <span class="op">=</span> <span class="st">"relabundance"</span><span class="op">)</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu">transformAssay</span><span class="op">(</span><span class="va">tse</span>, method <span class="op">=</span> <span class="st">"clr"</span>, pseudocount <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co"># Transform agglomerated data</span></span>
<span><span class="va">tse_family</span> <span class="op">&lt;-</span> <span class="fu">transformAssay</span><span class="op">(</span><span class="va">tse_family</span>, method <span class="op">=</span> <span class="st">"relabundance"</span><span class="op">)</span></span>
<span><span class="va">tse_family</span> <span class="op">&lt;-</span> <span class="fu">transformAssay</span><span class="op">(</span><span class="va">tse_family</span>, method <span class="op">=</span> <span class="st">"clr"</span>, pseudocount <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="analysis-of-community-diversity-and-composition-pande">Analysis of community diversity and composition (Pande)<a class="anchor" aria-label="anchor" href="#analysis-of-community-diversity-and-composition-pande"></a>
</h2>
<p>Idea for the story:</p>
<p>” After preprocessing the data, we can continue to analysis. It is
common to analyze the alpha diversity of samples. …</p>
<p>To analyze similarity between samples, we can perform PCoA. … ”</p>
</div>
<div class="section level2">
<h2 id="microbiome-data-integration-tuomas">Microbiome data integration (Tuomas)<a class="anchor" aria-label="anchor" href="#microbiome-data-integration-tuomas"></a>
</h2>
<p>As demonstrated, CRC patients appear to have altered microbiota.
However, this observation only indicates an association and does not
imply causality or reveal underlying mechanisms. To explore complex
mechanisms, we often need more comprehensive data. Efficiently handling
data from multiple experiments requires integrating these datasets into
a unified data container. Bioconductor offers two approaches for
this:</p>
<ul>
<li>Using the <code>altExp</code> slot within the <code>TreeSE</code>
object, which is a straightforward method.</li>
<li>Employing <code>MultiAssayExperiment</code>, a more general and
flexible approach, though it comes with a more complex data structure
compared to <code>TreeSE</code>.</li>
</ul>
<p>For detailed information on both approaches, refer to <span class="citation">[@sec-containers]</span>. In this tutorial, we will
focus on the simpler method using the <code>altExp</code> slot.</p>
<p>The data fetched at the beginning of this workflow is a
<code>list</code> containing multiple experiments, each represented as a
unique <code>(Tree)SE</code> object. Managing this list and the
connections between samples can be error-prone and cumbersome. To
simplify this, we can use the <code>altExp</code> slot to handle the
bookkeeping. Let’s start by checking which experiments are
available:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">data_list</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "pathway"  "taxonomy"</span></span></code></pre></div>
<p>As we see, we have metabolic pathway predictions in addition to
taxonomic profiling. We can add the entire list of experiments to the
<code>altExp</code> slot.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">altExps</span><span class="op">(</span><span class="va">tse</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">data_list</span></span>
<span><span class="va">tse</span></span>
<span><span class="co">#&gt; class: TreeSummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 304 60 </span></span>
<span><span class="co">#&gt; metadata(1): agglomerated_by_rank</span></span>
<span><span class="co">#&gt; assays(3): counts relabundance clr</span></span>
<span><span class="co">#&gt; rownames(304): [Bacteroides] pectinophilus [Clostridium] hylemonae ...</span></span>
<span><span class="co">#&gt;   Victivallis vadensis Weissella confusa</span></span>
<span><span class="co">#&gt; rowData names(7): superkingdom phylum ... genus species</span></span>
<span><span class="co">#&gt; colnames(60): GupDM_A_11 GupDM_A_15 ... GupDM_JO GupDM_JP</span></span>
<span><span class="co">#&gt; colData names(27): study_name subject_id ... disease_stage</span></span>
<span><span class="co">#&gt;   disease_location</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(2): pathway taxonomy</span></span>
<span><span class="co">#&gt; rowLinks: a LinkDataFrame (304 rows)</span></span>
<span><span class="co">#&gt; rowTree: 1 phylo tree(s) (10430 leaves)</span></span>
<span><span class="co">#&gt; colLinks: NULL</span></span>
<span><span class="co">#&gt; colTree: NULL</span></span></code></pre></div>
<p>With the <code>altExp</code> slot now populated, you can access the
data using the <code>altExp()</code> function and specify which
experiment you want to retrieve.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">altExp</span><span class="op">(</span><span class="va">tse</span>, <span class="st">"pathway"</span><span class="op">)</span></span>
<span><span class="co">#&gt; class: SummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 12535 60 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): pathway_abundance</span></span>
<span><span class="co">#&gt; rownames(12535): UNMAPPED UNINTEGRATED ... PWY-7197: pyrimidine</span></span>
<span><span class="co">#&gt;   deoxyribonucleotide</span></span>
<span><span class="co">#&gt;   phosphorylation|g__Desulfovibrio.s__Desulfovibrio_piger PWY-6703:</span></span>
<span><span class="co">#&gt;   preQ0 biosynthesis|g__Butyricimonas.s__Butyricimonas_virosa</span></span>
<span><span class="co">#&gt; rowData names(2): full_name pathway</span></span>
<span><span class="co">#&gt; colnames(60): GupDM_A_11 GupDM_A_15 ... GupDM_JO GupDM_JP</span></span>
<span><span class="co">#&gt; colData names(27): study_name subject_id ... disease_stage</span></span>
<span><span class="co">#&gt;   disease_location</span></span></code></pre></div>
<p>As mentioned earlier, the advantage of the <code>altExp</code> slot
is that it handles sample linkages for us. For example, if we select a
subset of data, the object automatically subsets all associated
experiments accordingly.</p>
<p>For example, let’s select only the samples from senior patients.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tse_sub</span> <span class="op">&lt;-</span> <span class="va">tse</span><span class="op">[</span> , <span class="va">tse</span><span class="op">$</span><span class="va">age_category</span> <span class="op">==</span> <span class="st">"senior"</span><span class="op">]</span></span>
<span><span class="va">tse_sub</span></span>
<span><span class="co">#&gt; class: TreeSummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 304 8 </span></span>
<span><span class="co">#&gt; metadata(1): agglomerated_by_rank</span></span>
<span><span class="co">#&gt; assays(3): counts relabundance clr</span></span>
<span><span class="co">#&gt; rownames(304): [Bacteroides] pectinophilus [Clostridium] hylemonae ...</span></span>
<span><span class="co">#&gt;   Victivallis vadensis Weissella confusa</span></span>
<span><span class="co">#&gt; rowData names(7): superkingdom phylum ... genus species</span></span>
<span><span class="co">#&gt; colnames(8): GupDM_A4 GupDM_A5 ... GupDM_JO GupDM_JP</span></span>
<span><span class="co">#&gt; colData names(27): study_name subject_id ... disease_stage</span></span>
<span><span class="co">#&gt;   disease_location</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(2): pathway taxonomy</span></span>
<span><span class="co">#&gt; rowLinks: a LinkDataFrame (304 rows)</span></span>
<span><span class="co">#&gt; rowTree: 1 phylo tree(s) (10430 leaves)</span></span>
<span><span class="co">#&gt; colLinks: NULL</span></span>
<span><span class="co">#&gt; colTree: NULL</span></span>
<span></span>
<span><span class="co"># Check that all sample names still match</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu">altExp</span><span class="op">(</span><span class="va">tse_sub</span>, <span class="st">"pathway"</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">tse_sub</span><span class="op">)</span> <span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<p>Now, we can see that the data has been successfully subsetted, with
all sample names matching across experiments.</p>
<p>In addition to adding the entire list, we can also add individual
experiments. Let’s add the agglomerated data separately to keep the
whole dataset organized.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">altExp</span><span class="op">(</span><span class="va">tse</span>, <span class="st">"family"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">tse_family</span></span>
<span><span class="va">tse</span></span>
<span><span class="co">#&gt; class: TreeSummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 304 60 </span></span>
<span><span class="co">#&gt; metadata(1): agglomerated_by_rank</span></span>
<span><span class="co">#&gt; assays(3): counts relabundance clr</span></span>
<span><span class="co">#&gt; rownames(304): [Bacteroides] pectinophilus [Clostridium] hylemonae ...</span></span>
<span><span class="co">#&gt;   Victivallis vadensis Weissella confusa</span></span>
<span><span class="co">#&gt; rowData names(7): superkingdom phylum ... genus species</span></span>
<span><span class="co">#&gt; colnames(60): GupDM_A_11 GupDM_A_15 ... GupDM_JO GupDM_JP</span></span>
<span><span class="co">#&gt; colData names(27): study_name subject_id ... disease_stage</span></span>
<span><span class="co">#&gt;   disease_location</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(3): pathway taxonomy family</span></span>
<span><span class="co">#&gt; rowLinks: a LinkDataFrame (304 rows)</span></span>
<span><span class="co">#&gt; rowTree: 1 phylo tree(s) (10430 leaves)</span></span>
<span><span class="co">#&gt; colLinks: NULL</span></span>
<span><span class="co">#&gt; colTree: NULL</span></span></code></pre></div>
<p>Now that the entire dataset is well-organized and in one place, let’s
dive into analyzing the interconnections between experiments.
Specifically, we’ll examine whether the abundance of certain bacteria is
associated with specific metabolic activity pathway.</p>
<p>We’ll perform a simple cross-association analysis to explore these
relationships. First, we’ll preprocess the functional annotation data by
agglomerating the pathways based on each pathway’s function, and then
further by prevalence.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Agglomerate based on pathways function</span></span>
<span><span class="fu">altExp</span><span class="op">(</span><span class="va">tse</span>, <span class="st">"pathway_func"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">agglomerateByVariable</span><span class="op">(</span></span>
<span>    <span class="fu">altExp</span><span class="op">(</span><span class="va">tse</span>, <span class="st">"pathway"</span><span class="op">)</span>,</span>
<span>    by <span class="op">=</span> <span class="st">"rows"</span>,</span>
<span>    f <span class="op">=</span> <span class="st">"pathway"</span></span>
<span>    <span class="op">)</span></span>
<span><span class="co"># Agglomerate based on prevalence</span></span>
<span><span class="fu">altExp</span><span class="op">(</span><span class="va">tse</span>, <span class="st">"pathway_prev"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">agglomerateByPrevalence</span><span class="op">(</span></span>
<span>    <span class="fu">altExp</span><span class="op">(</span><span class="va">tse</span>, <span class="st">"pathway_func"</span><span class="op">)</span>,</span>
<span>    assay.type <span class="op">=</span> <span class="st">"pathway_abundance"</span>,</span>
<span>    prevalence <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>    detection <span class="op">=</span> <span class="fl">0.001</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p>The data now includes the most prevalent pathways, each representing
a single metabolic activity. Next, we apply a transformation as the
final preprocessing step for our functional data.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">altExp</span><span class="op">(</span><span class="va">tse</span>, <span class="st">"pathway_prev"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">transformAssay</span><span class="op">(</span></span>
<span>    <span class="fu">altExp</span><span class="op">(</span><span class="va">tse</span>, <span class="st">"pathway_prev"</span><span class="op">)</span>,</span>
<span>    assay.type <span class="op">=</span> <span class="st">"pathway_abundance"</span>,</span>
<span>    method <span class="op">=</span> <span class="st">"clr"</span>,</span>
<span>    pseudocount <span class="op">=</span> <span class="fl">1</span></span>
<span>    <span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The assay contains only positive values. Applying a pseudocount may be</span></span>
<span><span class="co">#&gt; unnecessary.</span></span>
<span><span class="fu">altExp</span><span class="op">(</span><span class="va">tse</span>, <span class="st">"pathway_prev"</span><span class="op">)</span></span>
<span><span class="co">#&gt; class: SummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 24 60 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(2): pathway_abundance clr</span></span>
<span><span class="co">#&gt; rownames(24): BRANCHED-CHAIN-AA-SYN-PWY: superpathway of branched amino</span></span>
<span><span class="co">#&gt;   acid biosynthesis DTDPRHAMSYN-PWY: dTDP-L-rhamnose biosynthesis I ...</span></span>
<span><span class="co">#&gt;   VALSYN-PWY: L-valine biosynthesis Other</span></span>
<span><span class="co">#&gt; rowData names(2): full_name pathway</span></span>
<span><span class="co">#&gt; colnames(60): GupDM_A_11 GupDM_A_15 ... GupDM_JO GupDM_JP</span></span>
<span><span class="co">#&gt; colData names(27): study_name subject_id ... disease_stage</span></span>
<span><span class="co">#&gt;   disease_location</span></span></code></pre></div>
<p>We’re now ready to perform the cross-association analysis. We will
examine the associations between microbial families and the most
abundant pathways. To visualize the results, we’ll use a heatmap.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform cross-associaton</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">getCrossAssociation</span><span class="op">(</span></span>
<span>    <span class="va">tse</span>,</span>
<span>    altexp1 <span class="op">=</span> <span class="st">"family"</span>,</span>
<span>    altexp2 <span class="op">=</span> <span class="st">"pathway_prev"</span>,</span>
<span>    assay.type1 <span class="op">=</span> <span class="st">"clr"</span>,</span>
<span>    assay.type2 <span class="op">=</span> <span class="st">"clr"</span>,</span>
<span>    mode <span class="op">=</span> <span class="st">"matrix"</span>,</span>
<span>    test.signif <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>    <span class="op">)</span></span>
<span><span class="co">#&gt; Calculating correlations...</span></span>
<span><span class="co">#&gt; altexp1: family, altexp2: pathway_prev, assay.type1: clr, col.var1: -, assay.type2: clr, col.var2: -</span></span>
<span><span class="co">#&gt; by: 1, function: stats::cor.test, method: kendall, test.signif: TRUE, p.adj.method: fdr, paired: FALSE, show.warnings: TRUE</span></span>
<span><span class="co">#&gt; Converting table into matrices...</span></span>
<span></span>
<span></span>
<span><span class="co"># Function for marking significant correlations with "X"</span></span>
<span><span class="va">add_signif</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">j</span>, <span class="va">i</span>, <span class="va">x</span>, <span class="va">y</span>, <span class="va">width</span>, <span class="va">height</span>, <span class="va">fill</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># If the p-value is under threshold</span></span>
<span>    <span class="kw">if</span><span class="op">(</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">p_adj</span><span class="op">[</span><span class="va">i</span>, <span class="va">j</span><span class="op">]</span><span class="op">)</span> <span class="op">&amp;</span> <span class="va">res</span><span class="op">$</span><span class="va">p_adj</span><span class="op">[</span><span class="va">i</span>, <span class="va">j</span><span class="op">]</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">)</span><span class="op">{</span></span>
<span>        <span class="co"># Print "X"</span></span>
<span>        <span class="fu">grid.shadowtext</span><span class="op">(</span></span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%s"</span>, <span class="st">"X"</span><span class="op">)</span>, <span class="va">x</span>, <span class="va">y</span>, gp <span class="op">=</span> <span class="fu">gpar</span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">8</span>, col <span class="op">=</span> <span class="st">"#f5f5f5"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Create a heatmap</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">Heatmap</span><span class="op">(</span></span>
<span>    <span class="va">res</span><span class="op">$</span><span class="va">cor</span>,</span>
<span>    <span class="co"># Print values to cells</span></span>
<span>    cell_fun <span class="op">=</span> <span class="va">add_signif</span>,</span>
<span>    heatmap_legend_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>     title <span class="op">=</span> <span class="st">"correlation"</span>, legend_height <span class="op">=</span> <span class="fu">unit</span><span class="op">(</span><span class="fl">5</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    column_names_rot <span class="op">=</span> <span class="fl">45</span>,</span>
<span>    column_names_max_height <span class="op">=</span> <span class="fu">unit</span><span class="op">(</span><span class="fl">20</span>, <span class="st">"cm"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="eccb_workflow_files/figure-html/cross_association-1.png" width="1440"></p>
<p>Enterobacteriaceae shows a positive correlation with certain pathways
(indicated by x, which denotes statistical significance with an adjusted
p-value &lt; 0.05) that other microbial families are negatively
correlated with. For example, L-isoleucine biosynthesis is a pathway
that demonstrates this kind of association.</p>
<p>For more information on data integration, see the following
sections:</p>
<ul>
<li><span class="citation">[@sec-cross-correlation]</span></li>
<li><span class="citation">[@sec-multiassay_ordination]</span></li>
<li><span class="citation">[@sec-multi-omics-integration]</span></li>
</ul>
</div>
<div class="section level2">
<h2 id="microbial-network-analysis-stefanie">Microbial network analysis (Stefanie)<a class="anchor" aria-label="anchor" href="#microbial-network-analysis-stefanie"></a>
</h2>
<p>Idea for the story:</p>
<p>” We performed cross-association analysis in previous session.
Analyzing microbial networks, however, usually requires more
sophisticated apporoaches.</p>
<p>”</p>
</div>
<div class="section level2">
<h2 id="interactive-microbiome-data-exploration-with-iseetree-giulio">Interactive microbiome data exploration with iSEEtree (Giulio)<a class="anchor" aria-label="anchor" href="#interactive-microbiome-data-exploration-with-iseetree-giulio"></a>
</h2>
<p>Idea for the story:</p>
<p>” For interactive analysis, iSEEtree is available. It allows to
perform common analyses in graphical interface.</p>
<p>”</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Etc/UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    grid      stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">#&gt; [8] methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] shadowtext_0.1.4                scater_1.33.4                  </span></span>
<span><span class="co">#&gt;  [3] scuttle_1.15.4                  miaViz_1.13.10                 </span></span>
<span><span class="co">#&gt;  [5] ggraph_2.2.1                    ggplot2_3.5.1                  </span></span>
<span><span class="co">#&gt;  [7] mia_1.13.38                     MultiAssayExperiment_1.31.5    </span></span>
<span><span class="co">#&gt;  [9] curatedMetagenomicData_3.13.0   TreeSummarizedExperiment_2.13.0</span></span>
<span><span class="co">#&gt; [11] Biostrings_2.73.1               XVector_0.45.0                 </span></span>
<span><span class="co">#&gt; [13] SingleCellExperiment_1.27.2     SummarizedExperiment_1.35.1    </span></span>
<span><span class="co">#&gt; [15] Biobase_2.65.1                  GenomicRanges_1.57.1           </span></span>
<span><span class="co">#&gt; [17] GenomeInfoDb_1.41.1             IRanges_2.39.2                 </span></span>
<span><span class="co">#&gt; [19] S4Vectors_0.43.2                BiocGenerics_0.51.1            </span></span>
<span><span class="co">#&gt; [21] MatrixGenerics_1.17.0           matrixStats_1.4.1              </span></span>
<span><span class="co">#&gt; [23] ComplexHeatmap_2.21.0           BiocManager_1.30.25            </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] splines_4.4.1               ggplotify_0.1.2            </span></span>
<span><span class="co">#&gt;   [3] filelock_1.0.3              tibble_3.2.1               </span></span>
<span><span class="co">#&gt;   [5] polyclip_1.10-7             rpart_4.1.23               </span></span>
<span><span class="co">#&gt;   [7] DirichletMultinomial_1.47.0 lifecycle_1.0.4            </span></span>
<span><span class="co">#&gt;   [9] doParallel_1.0.17           lattice_0.22-6             </span></span>
<span><span class="co">#&gt;  [11] MASS_7.3-61                 SnowballC_0.7.1            </span></span>
<span><span class="co">#&gt;  [13] backports_1.5.0             magrittr_2.0.3             </span></span>
<span><span class="co">#&gt;  [15] Hmisc_5.1-3                 sass_0.4.9                 </span></span>
<span><span class="co">#&gt;  [17] rmarkdown_2.28              jquerylib_0.1.4            </span></span>
<span><span class="co">#&gt;  [19] yaml_2.3.10                 DBI_1.2.3                  </span></span>
<span><span class="co">#&gt;  [21] minqa_1.2.8                 RColorBrewer_1.1-3         </span></span>
<span><span class="co">#&gt;  [23] abind_1.4-5                 zlibbioc_1.51.1            </span></span>
<span><span class="co">#&gt;  [25] purrr_1.0.2                 yulab.utils_0.1.7          </span></span>
<span><span class="co">#&gt;  [27] nnet_7.3-19                 tweenr_2.0.3               </span></span>
<span><span class="co">#&gt;  [29] rappdirs_0.3.3              sandwich_3.1-0             </span></span>
<span><span class="co">#&gt;  [31] circlize_0.4.16             GenomeInfoDbData_1.2.12    </span></span>
<span><span class="co">#&gt;  [33] ggrepel_0.9.6               tokenizers_0.3.0           </span></span>
<span><span class="co">#&gt;  [35] irlba_2.3.5.1               tidytree_0.4.6             </span></span>
<span><span class="co">#&gt;  [37] vegan_2.6-8                 rbiom_1.0.3                </span></span>
<span><span class="co">#&gt;  [39] permute_0.9-7               pkgdown_2.1.0              </span></span>
<span><span class="co">#&gt;  [41] DelayedMatrixStats_1.27.3   codetools_0.2-20           </span></span>
<span><span class="co">#&gt;  [43] DelayedArray_0.31.11        ggforce_0.4.2              </span></span>
<span><span class="co">#&gt;  [45] tidyselect_1.2.1            shape_1.4.6.1              </span></span>
<span><span class="co">#&gt;  [47] aplot_0.2.3                 farver_2.1.2               </span></span>
<span><span class="co">#&gt;  [49] UCSC.utils_1.1.0            lme4_1.1-35.5              </span></span>
<span><span class="co">#&gt;  [51] ScaledMatrix_1.13.0         viridis_0.6.5              </span></span>
<span><span class="co">#&gt;  [53] BiocFileCache_2.13.0        base64enc_0.1-3            </span></span>
<span><span class="co">#&gt;  [55] jsonlite_1.8.8              GetoptLong_1.0.5           </span></span>
<span><span class="co">#&gt;  [57] BiocNeighbors_1.99.0        tidygraph_1.3.1            </span></span>
<span><span class="co">#&gt;  [59] decontam_1.25.0             Formula_1.2-5              </span></span>
<span><span class="co">#&gt;  [61] iterators_1.0.14            systemfonts_1.1.0          </span></span>
<span><span class="co">#&gt;  [63] foreach_1.5.2               ggnewscale_0.5.0           </span></span>
<span><span class="co">#&gt;  [65] tools_4.4.1                 treeio_1.29.1              </span></span>
<span><span class="co">#&gt;  [67] ragg_1.3.3                  Rcpp_1.0.13                </span></span>
<span><span class="co">#&gt;  [69] glue_1.7.0                  gridExtra_2.3              </span></span>
<span><span class="co">#&gt;  [71] SparseArray_1.5.34          mgcv_1.9-1                 </span></span>
<span><span class="co">#&gt;  [73] xfun_0.47                   dplyr_1.1.4                </span></span>
<span><span class="co">#&gt;  [75] withr_3.0.1                 fastmap_1.2.0              </span></span>
<span><span class="co">#&gt;  [77] boot_1.3-31                 bluster_1.15.1             </span></span>
<span><span class="co">#&gt;  [79] fansi_1.0.6                 digest_0.6.37              </span></span>
<span><span class="co">#&gt;  [81] rsvd_1.0.5                  mime_0.12                  </span></span>
<span><span class="co">#&gt;  [83] gridGraphics_0.5-1          R6_2.5.1                   </span></span>
<span><span class="co">#&gt;  [85] textshaping_0.4.0           colorspace_2.1-1           </span></span>
<span><span class="co">#&gt;  [87] Cairo_1.6-2                 lpSolve_5.6.20             </span></span>
<span><span class="co">#&gt;  [89] RSQLite_2.3.7               utf8_1.2.4                 </span></span>
<span><span class="co">#&gt;  [91] tidyr_1.3.1                 generics_0.1.3             </span></span>
<span><span class="co">#&gt;  [93] data.table_1.16.0           DECIPHER_3.1.4             </span></span>
<span><span class="co">#&gt;  [95] graphlayouts_1.1.1          httr_1.4.7                 </span></span>
<span><span class="co">#&gt;  [97] htmlwidgets_1.6.4           S4Arrays_1.5.7             </span></span>
<span><span class="co">#&gt;  [99] pkgconfig_2.0.3             gtable_0.3.5               </span></span>
<span><span class="co">#&gt; [101] blob_1.2.4                  janeaustenr_1.0.0          </span></span>
<span><span class="co">#&gt; [103] htmltools_0.5.8.1           clue_0.3-65                </span></span>
<span><span class="co">#&gt; [105] scales_1.3.0                png_0.1-8                  </span></span>
<span><span class="co">#&gt; [107] ggfun_0.1.6                 knitr_1.48                 </span></span>
<span><span class="co">#&gt; [109] rstudioapi_0.16.0           reshape2_1.4.4             </span></span>
<span><span class="co">#&gt; [111] rjson_0.2.22                checkmate_2.3.2            </span></span>
<span><span class="co">#&gt; [113] nlme_3.1-166                curl_5.2.2                 </span></span>
<span><span class="co">#&gt; [115] nloptr_2.1.1                zoo_1.8-12                 </span></span>
<span><span class="co">#&gt; [117] cachem_1.1.0                GlobalOptions_0.1.2        </span></span>
<span><span class="co">#&gt; [119] stringr_1.5.1               BiocVersion_3.20.0         </span></span>
<span><span class="co">#&gt; [121] parallel_4.4.1              vipor_0.4.7                </span></span>
<span><span class="co">#&gt; [123] foreign_0.8-87              AnnotationDbi_1.67.0       </span></span>
<span><span class="co">#&gt; [125] desc_1.4.3                  pillar_1.9.0               </span></span>
<span><span class="co">#&gt; [127] vctrs_0.6.5                 slam_0.1-53                </span></span>
<span><span class="co">#&gt; [129] BiocSingular_1.21.3         dbplyr_2.5.0               </span></span>
<span><span class="co">#&gt; [131] beachmat_2.21.6             cluster_2.1.6              </span></span>
<span><span class="co">#&gt; [133] beeswarm_0.4.0              htmlTable_2.4.3            </span></span>
<span><span class="co">#&gt; [135] evaluate_0.24.0             mvtnorm_1.3-1              </span></span>
<span><span class="co">#&gt; [137] cli_3.6.3                   compiler_4.4.1             </span></span>
<span><span class="co">#&gt; [139] rlang_1.1.4                 crayon_1.5.3               </span></span>
<span><span class="co">#&gt; [141] tidytext_0.4.2              mediation_4.5.0            </span></span>
<span><span class="co">#&gt; [143] plyr_1.8.9                  fs_1.6.4                   </span></span>
<span><span class="co">#&gt; [145] ggbeeswarm_0.7.2            stringi_1.8.4              </span></span>
<span><span class="co">#&gt; [147] viridisLite_0.4.2           BiocParallel_1.39.0        </span></span>
<span><span class="co">#&gt; [149] munsell_0.5.1               lazyeval_0.2.2             </span></span>
<span><span class="co">#&gt; [151] Matrix_1.7-0                ExperimentHub_2.13.1       </span></span>
<span><span class="co">#&gt; [153] patchwork_1.2.0             sparseMatrixStats_1.17.2   </span></span>
<span><span class="co">#&gt; [155] bit64_4.0.5                 KEGGREST_1.45.1            </span></span>
<span><span class="co">#&gt; [157] highr_0.11                  AnnotationHub_3.13.3       </span></span>
<span><span class="co">#&gt; [159] igraph_2.0.3                memoise_2.0.1              </span></span>
<span><span class="co">#&gt; [161] RcppParallel_5.1.9          bslib_0.8.0                </span></span>
<span><span class="co">#&gt; [163] ggtree_3.13.1               bit_4.0.5                  </span></span>
<span><span class="co">#&gt; [165] ape_5.8</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Tuomas Borman, Leo Lahti.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
