<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Orchestrating Microbiome Analysis with Bioconductor (ECCB) • miaDemo</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Orchestrating Microbiome Analysis with Bioconductor (ECCB)">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">miaDemo</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/eccb_workflow.html">Orchestrating Microbiome Analysis with Bioconductor (ECCB)</a>
    </li>
    <li>
      <a href="../articles/mia_demo.html">Orchestrating Microbiome Analysis with Bioconductor (EuroBioC2024)</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/microbiome/miaDemo" class="external-link">
    <span class="fa fa-github"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Orchestrating Microbiome Analysis with
Bioconductor (ECCB)</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/microbiome/miaDemo/blob/iseetree/vignettes/eccb_workflow.Rmd" class="external-link"><code>vignettes/eccb_workflow.Rmd</code></a></small>
      <div class="hidden name"><code>eccb_workflow.Rmd</code></div>

    </div>

    
    
<p>Authors: Tuomas Borman, Pande Erawijantari, Giulio Benedetti,
Stefanie Peschel, Leo Lahti. <br> Last modified: 16 September,
2024.</p>
<p><img src="figures/mia_logo.png" width="150"></p>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<div class="section level3">
<h3 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h3>
<p>This tutorial introduces the latest advances in Bioconductor tools
and data structures supporting microbiome analysis. We will show how
this can enhance interoperability across omics and walk through a
typical microbiome data science workflow. You will learn from active
developers how to access open microbiome data resources, utilize
optimized data structures, assess community diversity and composition,
integrate taxonomic and functional hierarchies, and visualize microbiome
data. We will follow the online book <a href="https://microbiome.github.io/OMA/docs/devel/" class="external-link">“Orchestrating
Microbiome Analysis with Bioconductor”</a> and Bioconductor
SummarizedExperiment framework, supporting optimized analysis and
integration of hierarchical, multi-domain microbiome data. Participants
are encouraged to install the latest versions of R and Bioconductor.</p>
</div>
<div class="section level3">
<h3 id="pre-requisites">Pre-requisites<a class="anchor" aria-label="anchor" href="#pre-requisites"></a>
</h3>
<ul>
<li>Basic knowledge of R coding</li>
<li>Understanding of microbiome research</li>
</ul>
<p>If your time allows, we recommend to spend some time to explore
beforehand <a href="https://microbiome.github.io/OMA/docs/devel/" class="external-link">Orchestrating
Microbiome Analysis (OMA) online book</a>.</p>
</div>
<div class="section level3">
<h3 id="participation">Participation<a class="anchor" aria-label="anchor" href="#participation"></a>
</h3>
<p>We have instructor-led demonstrations where participants are expected
to follow along. Questions are encouraged at any time during the
workshop.</p>
<p>It is preferred to use your own laptop. You can find help for
installation from <a href="https://microbiome.github.io/OMA/docs/devel/pages/training.html" class="external-link">here</a>.
There are also available <a href="https://workshop.bioconductor.org/" class="external-link">pre-installed virtual
machines</a> by Galaxy/Bioconductor. (Under “EUROBIOC2024” tab, session
called “Package Demo: mia”)</p>
</div>
<div class="section level3">
<h3 id="r-bioconductor-packages-used">
<em>R</em> / <em>Bioconductor</em> packages used<a class="anchor" aria-label="anchor" href="#r-bioconductor-packages-used"></a>
</h3>
<p>In this workshop, we will focus on <a href="https://bioconductor.org/packages/devel/bioc/html/mia.html" class="external-link">the
<em>mia</em> package</a>, which is designed for microbiome data science.
The <em>mia</em> package provides essential tools for analyzing
microbiome data using the <em>SummarizedExperiment</em> framework.</p>
<p>Additionally, we will cover <a href="https://github.com/stefpeschel/NetCoMi" class="external-link">NetCoMi</a> and <a href="https://www.bioconductor.org/packages/devel/bioc/html/iSEEtree.html" class="external-link">iSEEtree</a>
in network and interactive data exploration session, respectively.</p>
</div>
<div class="section level3">
<h3 id="time-outline">Time outline<a class="anchor" aria-label="anchor" href="#time-outline"></a>
</h3>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Bioconductor resources for microbiome data science</td>
<td>30m</td>
</tr>
<tr class="even">
<td>Analysis of community diversity and composition</td>
<td>30m</td>
</tr>
<tr class="odd">
<td>Microbiome data integration</td>
<td>30m</td>
</tr>
<tr class="even">
<td>Coffee break</td>
<td>30m</td>
</tr>
<tr class="odd">
<td>Microbial network analysis</td>
<td>30m</td>
</tr>
<tr class="even">
<td>Interactive microbiome data exploration with iSEEtree</td>
<td>30m</td>
</tr>
<tr class="odd">
<td>Recap and Q&amp;A</td>
<td>30m</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="workshop-goals-and-objectives">Workshop goals and objectives<a class="anchor" aria-label="anchor" href="#workshop-goals-and-objectives"></a>
</h3>
<div class="section level4">
<h4 id="learning-goals">Learning goals<a class="anchor" aria-label="anchor" href="#learning-goals"></a>
</h4>
<ul>
<li><p><strong>Methods for microbiome analysis</strong>: Learn to
harness <em>SummarizedExperiment</em> ecosystem in your
project.</p></li>
<li><p><strong>Data structure</strong>: Understand how to utilize the
<em>TreeSummarizedExperiment</em> class for effective microbiome data
analysis.</p></li>
<li><p><strong>Access resources</strong>: Get familiar with additional
tools and resources, including <a href="https://microbiome.github.io/OMA/docs/devel" class="external-link">the OMA online
book</a>.</p></li>
</ul>
</div>
<div class="section level4">
<h4 id="learning-objectives">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h4>
<ul>
<li><p><strong>Analyze and apply methods</strong>: Apply the framework
to process and analyze microbiome data.</p></li>
<li><p><strong>Visualize data</strong>: Generate and interpret common
graphs for microbiome data</p></li>
<li><p><strong>Explore documentation</strong>: Use the <a href="https://microbiome.github.io/OMA/docs/devel" class="external-link">OMA</a> to explore
additional tools and methods.</p></li>
</ul>
</div>
</div>
</div>
<div class="section level2">
<h2 id="workshop">Workshop<a class="anchor" aria-label="anchor" href="#workshop"></a>
</h2>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In this workflow, we demonstrate how to analyze data from publicly
available microbiome resources using Bioconductor methods. We will cover
essential data wrangling steps, explore alpha and beta diversity
metrics, and introduce the fundamentals of data integration and
microbial network analysis. Additionally, we will explore how to utilize
the tools through a graphical interface.</p>
</div>
<div class="section level2">
<h2 id="bioconductor-resources-for-microbiome-data-science">Bioconductor resources for microbiome data science<a class="anchor" aria-label="anchor" href="#bioconductor-resources-for-microbiome-data-science"></a>
</h2>
<div class="section level3">
<h3 id="data-container">Data container<a class="anchor" aria-label="anchor" href="#data-container"></a>
</h3>
<p>The <em>miaverse</em> framework is based on
<em>TreeSummarizedExperiment</em> data container, extension to common
<em>SummarizedExperiment</em> object.</p>
<div class="figure" style="text-align: center">
<img src="figures/TreeSE.png" alt="_TreeSummarizedExperiment_ class" width="80%"><p class="caption">
<em>TreeSummarizedExperiment</em> class
</p>
</div>
</div>
<div class="section level3">
<h3 id="load-packages">Load packages<a class="anchor" aria-label="anchor" href="#load-packages"></a>
</h3>
<p>First, we load the required packages into the session. The following
script ensures that any packages not already installed are automatically
installed.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># List of packages that we need</span></span>
<span><span class="va">packages</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"ComplexHeatmap"</span>,</span>
<span>    <span class="st">"curatedMetagenomicData"</span>,</span>
<span>    <span class="st">"dplyr"</span>,</span>
<span>    <span class="st">"ggsignif"</span>,</span>
<span>    <span class="st">"iSEEtree"</span>,</span>
<span>    <span class="st">"mia"</span>,</span>
<span>    <span class="st">"miaViz"</span>,</span>
<span>    <span class="st">"NetCoMi"</span>,</span>
<span>    <span class="st">"scater"</span>,</span>
<span>    <span class="st">"shadowtext"</span>,</span>
<span>    <span class="st">"patchwork"</span>,</span>
<span>    <span class="st">"scater"</span>,</span>
<span>    <span class="st">"SpiecEasi"</span>,</span>
<span>    <span class="st">"tidyr"</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="co"># Get packages that are already installed</span></span>
<span><span class="va">packages_already_installed</span> <span class="op">&lt;-</span> <span class="va">packages</span><span class="op">[</span> <span class="va">packages</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span> <span class="op">]</span></span>
<span><span class="co"># Get packages that need to be installed</span></span>
<span><span class="va">packages_need_to_install</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">setdiff</a></span><span class="op">(</span> <span class="va">packages</span>, <span class="va">packages_already_installed</span> <span class="op">)</span></span>
<span></span>
<span><span class="co"># Install GitHub packages</span></span>
<span><span class="kw">if</span><span class="op">(</span> <span class="st">"SpiecEasi"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">packages_need_to_install</span> <span class="op">)</span><span class="op">{</span></span>
<span>    <span class="co"># SpiecEasi must be installed from "latentcor" branch</span></span>
<span>    <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"zdk123/SpiecEasi"</span>, force <span class="op">=</span> <span class="cn">TRUE</span>, ref <span class="op">=</span> <span class="st">"latentcor"</span><span class="op">)</span></span>
<span>    <span class="co"># Remove from installation list</span></span>
<span>    <span class="va">packages_need_to_install</span> <span class="op">&lt;-</span> <span class="va">packages_need_to_install</span><span class="op">[</span> <span class="op">!</span><span class="va">packages_need_to_install</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="st">"SpiecEasi"</span> <span class="op">]</span></span>
<span><span class="op">}</span></span>
<span><span class="kw">if</span><span class="op">(</span> <span class="st">"SPRING"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">packages_need_to_install</span> <span class="op">)</span><span class="op">{</span></span>
<span>    <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"GraceYoon/SPRING"</span><span class="op">)</span></span>
<span>    <span class="co"># Remove from installation list</span></span>
<span>    <span class="va">packages_need_to_install</span> <span class="op">&lt;-</span> <span class="va">packages_need_to_install</span><span class="op">[</span> <span class="op">!</span><span class="va">packages_need_to_install</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="st">"SPRING"</span> <span class="op">]</span></span>
<span><span class="op">}</span></span>
<span><span class="kw">if</span><span class="op">(</span> <span class="st">"NetCoMi"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">packages_need_to_install</span> <span class="op">)</span><span class="op">{</span></span>
<span>    <span class="co"># NetCoMi must be installed from "TSE" branch</span></span>
<span>    <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span></span>
<span>        <span class="st">"stefpeschel/NetCoMi"</span>, force <span class="op">=</span> <span class="cn">TRUE</span>, ref <span class="op">=</span> <span class="st">"TSE"</span>,</span>
<span>        dependencies <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Depends"</span>, <span class="st">"Imports"</span>, <span class="st">"LinkingTo"</span><span class="op">)</span>,</span>
<span>        repos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"https://cloud.r-project.org/"</span>, <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/repositories.html" class="external-link">repositories</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="co"># Remove from installation list</span></span>
<span>    <span class="va">packages_need_to_install</span> <span class="op">&lt;-</span> <span class="va">packages_need_to_install</span><span class="op">[</span> <span class="op">!</span><span class="va">packages_need_to_install</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="st">"NetCoMi"</span> <span class="op">]</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Loads BiocManager into the session. Install it if it is not already installed.</span></span>
<span><span class="kw">if</span><span class="op">(</span> <span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span><span class="op">)</span> <span class="op">)</span><span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># If there are packages that need to be installed, installs them with BiocManager</span></span>
<span><span class="co"># Updates old packages.</span></span>
<span><span class="kw">if</span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">packages_need_to_install</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span> <span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span>version <span class="op">=</span> <span class="st">"devel"</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="va">packages_need_to_install</span>, ask <span class="op">=</span> <span class="cn">FALSE</span>, update <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Load all packages into session. Stop if there are packages that were not</span></span>
<span><span class="co"># successfully loaded</span></span>
<span><span class="va">pkgs_not_loaded</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">packages</span>, <span class="va">require</span>, character.only <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">pkgs_not_loaded</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">pkgs_not_loaded</span><span class="op">)</span><span class="op">[</span> <span class="va">pkgs_not_loaded</span> <span class="op">]</span></span>
<span><span class="kw">if</span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">pkgs_not_loaded</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span> <span class="op">)</span><span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span></span>
<span>        <span class="st">"Error in loading the following packages into the session: '"</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">pkgs_not_loaded</span>, collapse <span class="op">=</span> <span class="st">"', '"</span><span class="op">)</span>, <span class="st">"'"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>If you encountered an error, install packages manually.</p>
</div>
<div class="section level3">
<h3 id="importing-data">Importing data<a class="anchor" aria-label="anchor" href="#importing-data"></a>
</h3>
<p>There are several openly available datasets listed in <a href="https://microbiome.github.io/OMA/docs/devel/pages/import.html#sec-example-data" class="external-link">OMA</a>.
In this workflow, we will retrieve data from the curatedMetagenomicData
resource. We’ll focus on a dataset that includes both colorectal cancer
(CRC) patients and control subjects. Our goal is to investigate whether
CRC is associated with alterations in the gut microbiota.</p>
<p>Let’s fetch the data from the database.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Study from which we want to fetch data</span></span>
<span><span class="va">study</span> <span class="op">&lt;-</span> <span class="st">"GuptaA_2019"</span></span>
<span><span class="co"># Experiments that we want to fetch</span></span>
<span><span class="va">experiments</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pathway_abundance"</span>, <span class="st">"relative_abundance"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span>collapse <span class="op">=</span> <span class="st">"|"</span><span class="op">)</span></span>
<span><span class="va">search_pattern</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">study</span>, <span class="st">".("</span>, <span class="va">experiments</span>, <span class="st">")"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Fetch data</span></span>
<span><span class="va">data_list</span> <span class="op">&lt;-</span> <span class="fu">curatedMetagenomicData</span><span class="op">(</span></span>
<span>    <span class="va">search_pattern</span>, dryrun <span class="op">=</span> <span class="cn">FALSE</span>, counts <span class="op">=</span> <span class="cn">TRUE</span>, rownames <span class="op">=</span> <span class="st">"short"</span><span class="op">)</span></span>
<span><span class="co">#&gt; $`2021-03-31.GuptaA_2019.relative_abundance`</span></span>
<span><span class="co">#&gt; dropping rows without rowTree matches:</span></span>
<span><span class="co">#&gt;   k__Bacteria|p__Actinobacteria|c__Coriobacteriia|o__Coriobacteriales|f__Coriobacteriaceae|g__Collinsella|s__Collinsella_stercoris</span></span>
<span><span class="co">#&gt;   k__Bacteria|p__Actinobacteria|c__Coriobacteriia|o__Coriobacteriales|f__Coriobacteriaceae|g__Enorma|s__[Collinsella]_massiliensis</span></span>
<span><span class="co">#&gt;   k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Carnobacteriaceae|g__Granulicatella|s__Granulicatella_elegans</span></span>
<span><span class="co">#&gt;   k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Ruminococcaceae|g__Ruminococcus|s__Ruminococcus_champanellensis</span></span>
<span><span class="co">#&gt;   k__Bacteria|p__Proteobacteria|c__Betaproteobacteria|o__Burkholderiales|f__Sutterellaceae|g__Sutterella|s__Sutterella_parvirubra</span></span>
<span></span>
<span><span class="co"># Add new names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">data_list</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pathway"</span>, <span class="st">"taxonomy"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># The assay name for taxonomy profile is incorrect since we have counts, not</span></span>
<span><span class="co"># relative abundances</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"taxonomy"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu">assayNames</span><span class="op">(</span><span class="va">tse</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"counts"</span></span>
<span><span class="va">data_list</span><span class="op">[[</span><span class="st">"taxonomy"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">tse</span></span>
<span></span>
<span><span class="co"># Rename also pathway abundance. They are in relative scale</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"pathway"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu">assayNames</span><span class="op">(</span><span class="va">tse</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"relabundance"</span></span>
<span><span class="va">data_list</span><span class="op">[[</span><span class="st">"pathway"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">tse</span></span>
<span></span>
<span><span class="co"># For pathway data, extract pathway information from rownames and add them to</span></span>
<span><span class="co"># rowData</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"pathway"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu">rowData</span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">[[</span><span class="st">"full_name"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span>
<span><span class="fu">rowData</span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">[[</span><span class="st">"pathway"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"\\|.*"</span>, <span class="st">""</span>, <span class="fu">rowData</span><span class="op">(</span><span class="va">tse</span><span class="op">)</span><span class="op">[[</span><span class="st">"full_name"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">data_list</span><span class="op">[[</span><span class="st">"pathway"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">tse</span></span>
<span></span>
<span><span class="co"># Get taxonomy object</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"taxonomy"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">tse</span></span>
<span><span class="co">#&gt; class: TreeSummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 304 60 </span></span>
<span><span class="co">#&gt; metadata(1): agglomerated_by_rank</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(304): [Bacteroides] pectinophilus [Clostridium] hylemonae ...</span></span>
<span><span class="co">#&gt;   Victivallis vadensis Weissella confusa</span></span>
<span><span class="co">#&gt; rowData names(7): superkingdom phylum ... genus species</span></span>
<span><span class="co">#&gt; colnames(60): GupDM_A_11 GupDM_A_15 ... GupDM_JO GupDM_JP</span></span>
<span><span class="co">#&gt; colData names(27): study_name subject_id ... disease_stage</span></span>
<span><span class="co">#&gt;   disease_location</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; rowLinks: a LinkDataFrame (304 rows)</span></span>
<span><span class="co">#&gt; rowTree: 1 phylo tree(s) (10430 leaves)</span></span>
<span><span class="co">#&gt; colLinks: NULL</span></span>
<span><span class="co">#&gt; colTree: NULL</span></span></code></pre></div>
<p>The output is a <code>TreeSummarizedExperiment</code>
(<code>TreeSE</code>) object. For details on how to handle and access
the data, refer to <a href="https://microbiome.github.io/OMA/docs/devel/pages/containers.html" class="external-link">data
containers chapter in OMA</a>.</p>
</div>
<div class="section level3">
<h3 id="preprocess">Preprocess<a class="anchor" aria-label="anchor" href="#preprocess"></a>
</h3>
<p>A common step in microbiome workflows is agglomeration, where we
summarize data at specific taxonomy ranks. Additionally, preprocessing
involves transforming the data to address the unique characteristics of
microbiome profiling data, for instance. Details on agglomeration and
transformation are covered in <a href="https://microbiome.github.io/OMA/docs/devel/pages/agglomeration.html" class="external-link">agglomeration</a>
and <a href="https://microbiome.github.io/OMA/docs/devel/pages/transformation.html" class="external-link">transformation</a>
chapters in OMA, respectively.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Agglomerate data</span></span>
<span><span class="va">tse_family</span> <span class="op">&lt;-</span> <span class="fu">agglomerateByRank</span><span class="op">(</span><span class="va">tse</span>, rank <span class="op">=</span> <span class="st">"family"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The following values are already present in `metadata` and will be</span></span>
<span><span class="co">#&gt; overwritten: 'agglomerated_by_rank'. Consider using the 'name' argument to</span></span>
<span><span class="co">#&gt; specify alternative names.</span></span>
<span></span>
<span><span class="co"># Transform the main TreeSE</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu">transformAssay</span><span class="op">(</span><span class="va">tse</span>, method <span class="op">=</span> <span class="st">"relabundance"</span><span class="op">)</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu">transformAssay</span><span class="op">(</span><span class="va">tse</span>, method <span class="op">=</span> <span class="st">"clr"</span>, pseudocount <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co"># Transform agglomerated data</span></span>
<span><span class="va">tse_family</span> <span class="op">&lt;-</span> <span class="fu">transformAssay</span><span class="op">(</span><span class="va">tse_family</span>, method <span class="op">=</span> <span class="st">"relabundance"</span><span class="op">)</span></span>
<span><span class="va">tse_family</span> <span class="op">&lt;-</span> <span class="fu">transformAssay</span><span class="op">(</span><span class="va">tse_family</span>, method <span class="op">=</span> <span class="st">"clr"</span>, pseudocount <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="analysis-of-community-diversity-and-composition">Analysis of community diversity and composition<a class="anchor" aria-label="anchor" href="#analysis-of-community-diversity-and-composition"></a>
</h2>
<p>Now that the data has been pre-processed, we could start to analyze
the overview of the microbiome profile. For the dataset above, we can
continue to perform analysis, for example to answer these questions:</p>
<ol style="list-style-type: decimal">
<li><p>How does the microbial community differ between the CRC (n=30) vs
healthy group (n=30)?</p></li>
<li><p>How does the beta diversity compare between the CRC (n=30) vs
healthy group (n=30)?</p></li>
<li><p>What species shows significantly different relative abundance
between 2 groups?</p></li>
</ol>
<p>Let’s start with analyzing the differences on microbial community
diversity between CRC and healthy control group. Community diversity is
a central concept in microbiome research, and in the human microbiome
study it has been shown that the microbial diversity associated with the
host health status. Several diversity indices are available in the
ecological literature. You can explore further <a href="https://microbiome.github.io/OMA/docs/devel/pages/alpha_diversity.html" class="external-link">in
the OMA book</a> about what analysis in-term of community diversity can
be analyzed using mia package.</p>
<p>For simplicity, let’s calculate the microbial richness and alpha
diversity using the Shannon index (other calculation using different
index also available, and you can explore further)</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu">addAlpha</span><span class="op">(</span><span class="va">tse</span>, index <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"observed"</span>, <span class="st">"shannon"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># to see the observed and shannon calculation</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tse</span><span class="op">$</span><span class="va">observed</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1]  55  29 148  68  63  51</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tse</span><span class="op">$</span><span class="va">shannon</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 2.755092 1.044521 3.305171 2.744079 2.570301 2.937390</span></span></code></pre></div>
<p>To see the differences between healthy and CRC group, let’s visualize
the species richness and shannon diversity using violin plot (you can
also experiment with boxplot). Additionally, let’s also calculate the
statistical significance and plot it together using ggsignif with the
wilcoxon test. Please notes that the package does not support p-value
adjustment (such as with “fdr”). For more complex test, please see <a href="https://microbiome.github.io/OMA/docs/devel/pages/alpha_diversity.html#visualizing-significance-in-group-wise-comparisons" class="external-link">example
here</a> on how you can calculate p-value with adjustment using
<code>stat_compare_means</code> function from the ggpubr package, and
add visually appealing p-values to the plots or if you want to perform
more than two groups comparison.</p>
<p>The information about diseases state is available from the
<code>colData</code> of <code>tse</code> object, in the “diseases”
column.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Changes old levels with new levels</span></span>
<span><span class="va">tse</span><span class="op">$</span><span class="va">disease</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">tse</span><span class="op">$</span><span class="va">disease</span><span class="op">)</span></span>
<span><span class="co"># Access the levels of the factor</span></span>
<span><span class="va">disease_levels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">tse</span><span class="op">$</span><span class="va">disease</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Compute all pairwise combinations of these levels</span></span>
<span><span class="va">comb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/combn.html" class="external-link">combn</a></span><span class="op">(</span><span class="va">disease_levels</span>, <span class="fl">2</span>, simplify <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Observed richness plot</span></span>
<span><span class="va">observed_plot</span> <span class="op">&lt;-</span> <span class="fu">plotColData</span><span class="op">(</span></span>
<span>    <span class="va">tse</span>, </span>
<span>    y <span class="op">=</span> <span class="st">"observed"</span>, </span>
<span>    x <span class="op">=</span> <span class="st">"disease"</span>,</span>
<span>    colour_by <span class="op">=</span> <span class="st">"disease"</span></span>
<span>    <span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_signif</span><span class="op">(</span>comparisons <span class="op">=</span> <span class="va">comb</span>, map_signif_level <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">theme</span><span class="op">(</span>text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Shannon diversity plot</span></span>
<span><span class="va">shannon_plot</span> <span class="op">&lt;-</span> <span class="fu">plotColData</span><span class="op">(</span></span>
<span>    <span class="va">tse</span>, </span>
<span>    y <span class="op">=</span> <span class="st">"shannon"</span>, </span>
<span>    x <span class="op">=</span> <span class="st">"disease"</span>,</span>
<span>    colour_by <span class="op">=</span> <span class="st">"disease"</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_signif</span><span class="op">(</span>comparisons <span class="op">=</span> <span class="va">comb</span>, map_signif_level <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">theme</span><span class="op">(</span>text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Combine the two plots into a grid</span></span>
<span><span class="va">combined_plot</span> <span class="op">&lt;-</span> <span class="va">observed_plot</span> <span class="op">+</span> <span class="va">shannon_plot</span> <span class="op">+</span> <span class="fu">plot_layout</span><span class="op">(</span>guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Display the combined plot</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">combined_plot</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in wilcox.test.default(c(55, 29, 148, 68, 63, 51, 37, 44, 45, 61, :</span></span>
<span><span class="co">#&gt; cannot compute exact p-value with ties</span></span></code></pre></div>
<p><img src="eccb_workflow_files/figure-html/alpha%20diversity%20viz-1.png" width="700"></p>
<p><strong>Let’s pause and think</strong></p>
<ul>
<li><p>Can you also do the diversity analysis using different
indexes?</p></li>
<li><p>What can you inffer by these observations?</p></li>
</ul>
<p>Let’s continue observing the beta-diversity between 2 groups. Beta
diversity quantifies the dissimilarity between communities (multiple
samples), as opposed to alpha diversity, which focuses on variation
within a community (one sample). This <a href="https://microbiome.github.io/OMA/docs/devel/pages/beta_diversity.html" class="external-link">OMA
book section</a> explained various indexes that commonly used in the
microbial ecology context.</p>
<p>For this tutorial, let’s try to calculate the bray-curtis distance
and plot the PCoA.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform PCoA</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu">runMDS</span><span class="op">(</span></span>
<span>    <span class="va">tse</span>,</span>
<span>    FUN <span class="op">=</span> <span class="va">getDissimilarity</span>,</span>
<span>    method <span class="op">=</span> <span class="st">"bray"</span>,</span>
<span>    assay.type <span class="op">=</span> <span class="st">"relabundance"</span>,</span>
<span>    name <span class="op">=</span> <span class="st">"MDS_bray"</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="co"># Create ggplot object</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">plotReducedDim</span><span class="op">(</span><span class="va">tse</span>, <span class="st">"MDS_bray"</span>, colour_by <span class="op">=</span> <span class="st">"disease"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate explained variance</span></span>
<span><span class="va">e</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="fu">reducedDim</span><span class="op">(</span><span class="va">tse</span>, <span class="st">"MDS_bray"</span><span class="op">)</span>, <span class="st">"eig"</span><span class="op">)</span></span>
<span><span class="va">rel_eig</span> <span class="op">&lt;-</span> <span class="va">e</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">e</span><span class="op">[</span><span class="va">e</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add explained variance for each axis</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span> <span class="fu">labs</span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"PCoA 1 ("</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fl">100</span> <span class="op">*</span> <span class="va">rel_eig</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="fl">1</span><span class="op">)</span>, <span class="st">"%"</span>, <span class="st">")"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span>,</span>
<span>    y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"PCoA 2 ("</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fl">100</span> <span class="op">*</span> <span class="va">rel_eig</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="fl">1</span><span class="op">)</span>, <span class="st">"%"</span>, <span class="st">")"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="va">p1</span></span></code></pre></div>
<p><img src="eccb_workflow_files/figure-html/PCoA-1.png" width="700"></p>
<p>Sometimes, we also want to create ellipse to guide our visualization
whether the microbial community in two groups is different. You can
achieve it using the code below.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p_ellipse</span> <span class="op">&lt;-</span> <span class="va">p1</span> <span class="op">+</span> <span class="fu">stat_ellipse</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>color <span class="op">=</span> <span class="va">colour_by</span><span class="op">)</span>, level <span class="op">=</span> <span class="fl">0.95</span><span class="op">)</span></span>
<span><span class="va">p_ellipse</span></span></code></pre></div>
<p><img src="eccb_workflow_files/figure-html/PCoA%20elipse-1.png" width="700"></p>
<p>Additionally, we may want to also perform PERMANOVA analysis to
further estimate whether the microbial community is statistically
different. Please explore the comprehensive example in the <a href="https://microbiome.github.io/OMA/docs/devel/pages/beta_diversity.html#redundancy-analysis" class="external-link">OMA
book</a>.</p>
<p><strong>Let’s pause and think</strong></p>
<ul>
<li><p>Above is the example for unsupervised ordination, can you give
example of it supervised counterpart?</p></li>
<li><p>Can you distinguish between both method and when to use
it?</p></li>
</ul>
<p>OMA book also provide further exploration to explore the microbial
community structure, for example based on <a href="https://microbiome.github.io/OMA/docs/devel/pages/clustering.html" class="external-link">community
typing</a>.</p>
<p>Another primary analysis that usually can be done in microbiome study
is <strong>differential abundance analysis (DAA)</strong> for example to
answer the research question 3. However, due to limited time slots, we
will not demonstrate in this tutorial. You can explore various example
in OMA book for <a href="https://microbiome.github.io/OMA/docs/devel/pages/differential_abundance.html" class="external-link">DAA
here</a></p>
</div>
<div class="section level2">
<h2 id="microbiome-data-integration">Microbiome data integration<a class="anchor" aria-label="anchor" href="#microbiome-data-integration"></a>
</h2>
<div class="section level3">
<h3 id="data-handling">Data handling<a class="anchor" aria-label="anchor" href="#data-handling"></a>
</h3>
<p>As demonstrated, CRC patients appear to have altered microbiota.
However, this observation only indicates an association and does not
imply causality or reveal underlying mechanisms. To explore complex
mechanisms, we often need more comprehensive data. Efficiently handling
data from multiple experiments requires integrating these datasets into
a unified data container. Bioconductor offers two approaches for
this:</p>
<ul>
<li>Using the <code>altExp</code> slot within the <code>TreeSE</code>
object, which is a straightforward method.</li>
<li>Employing <code>MultiAssayExperiment</code>, a more general and
flexible approach, though it comes with a more complex data structure
compared to <code>TreeSE</code>.</li>
</ul>
<p>For detailed information on both approaches, refer to <a href="https://microbiome.github.io/OMA/docs/devel/pages/containers.html" class="external-link">OMA</a>.
In this tutorial, we will focus on the simpler method using the
<code>altExp</code> slot.</p>
<p>The data fetched at the beginning of this workflow is a
<code>list</code> containing multiple experiments, each represented as a
unique <code>(Tree)SE</code> object. Managing this list and the
connections between samples can be error-prone and cumbersome. To
simplify this, we can use the <code>altExp</code> slot to handle the
bookkeeping. Let’s start by checking which experiments are
available:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">data_list</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "pathway"  "taxonomy"</span></span></code></pre></div>
<p>As we see, we have metabolic pathway predictions in addition to
taxonomic profiling. We can add the entire list of experiments to the
<code>altExp</code> slot.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">altExps</span><span class="op">(</span><span class="va">tse</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">data_list</span></span>
<span><span class="va">tse</span></span>
<span><span class="co">#&gt; class: TreeSummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 304 60 </span></span>
<span><span class="co">#&gt; metadata(1): agglomerated_by_rank</span></span>
<span><span class="co">#&gt; assays(3): counts relabundance clr</span></span>
<span><span class="co">#&gt; rownames(304): [Bacteroides] pectinophilus [Clostridium] hylemonae ...</span></span>
<span><span class="co">#&gt;   Victivallis vadensis Weissella confusa</span></span>
<span><span class="co">#&gt; rowData names(7): superkingdom phylum ... genus species</span></span>
<span><span class="co">#&gt; colnames(60): GupDM_A_11 GupDM_A_15 ... GupDM_JO GupDM_JP</span></span>
<span><span class="co">#&gt; colData names(29): study_name subject_id ... observed shannon</span></span>
<span><span class="co">#&gt; reducedDimNames(1): MDS_bray</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(2): pathway taxonomy</span></span>
<span><span class="co">#&gt; rowLinks: a LinkDataFrame (304 rows)</span></span>
<span><span class="co">#&gt; rowTree: 1 phylo tree(s) (10430 leaves)</span></span>
<span><span class="co">#&gt; colLinks: NULL</span></span>
<span><span class="co">#&gt; colTree: NULL</span></span></code></pre></div>
<p>With the <code>altExp</code> slot now populated, you can access the
data using the <code>altExp()</code> function and specify which
experiment you want to retrieve.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">altExp</span><span class="op">(</span><span class="va">tse</span>, <span class="st">"pathway"</span><span class="op">)</span></span>
<span><span class="co">#&gt; class: SummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 12535 60 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): relabundance</span></span>
<span><span class="co">#&gt; rownames(12535): UNMAPPED UNINTEGRATED ... PWY-7197: pyrimidine</span></span>
<span><span class="co">#&gt;   deoxyribonucleotide</span></span>
<span><span class="co">#&gt;   phosphorylation|g__Desulfovibrio.s__Desulfovibrio_piger PWY-6703:</span></span>
<span><span class="co">#&gt;   preQ0 biosynthesis|g__Butyricimonas.s__Butyricimonas_virosa</span></span>
<span><span class="co">#&gt; rowData names(2): full_name pathway</span></span>
<span><span class="co">#&gt; colnames(60): GupDM_A_11 GupDM_A_15 ... GupDM_JO GupDM_JP</span></span>
<span><span class="co">#&gt; colData names(27): study_name subject_id ... disease_stage</span></span>
<span><span class="co">#&gt;   disease_location</span></span></code></pre></div>
<p>As mentioned earlier, the advantage of the <code>altExp</code> slot
is that it handles sample linkages for us. For example, if we select a
subset of data, the object automatically subsets all associated
experiments accordingly.</p>
<p>For example, let’s select only the samples from senior patients.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tse_sub</span> <span class="op">&lt;-</span> <span class="va">tse</span><span class="op">[</span> , <span class="va">tse</span><span class="op">$</span><span class="va">age_category</span> <span class="op">==</span> <span class="st">"senior"</span><span class="op">]</span></span>
<span><span class="va">tse_sub</span></span>
<span><span class="co">#&gt; class: TreeSummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 304 8 </span></span>
<span><span class="co">#&gt; metadata(1): agglomerated_by_rank</span></span>
<span><span class="co">#&gt; assays(3): counts relabundance clr</span></span>
<span><span class="co">#&gt; rownames(304): [Bacteroides] pectinophilus [Clostridium] hylemonae ...</span></span>
<span><span class="co">#&gt;   Victivallis vadensis Weissella confusa</span></span>
<span><span class="co">#&gt; rowData names(7): superkingdom phylum ... genus species</span></span>
<span><span class="co">#&gt; colnames(8): GupDM_A4 GupDM_A5 ... GupDM_JO GupDM_JP</span></span>
<span><span class="co">#&gt; colData names(29): study_name subject_id ... observed shannon</span></span>
<span><span class="co">#&gt; reducedDimNames(1): MDS_bray</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(2): pathway taxonomy</span></span>
<span><span class="co">#&gt; rowLinks: a LinkDataFrame (304 rows)</span></span>
<span><span class="co">#&gt; rowTree: 1 phylo tree(s) (10430 leaves)</span></span>
<span><span class="co">#&gt; colLinks: NULL</span></span>
<span><span class="co">#&gt; colTree: NULL</span></span>
<span></span>
<span><span class="co"># Check that all sample names still match</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu">altExp</span><span class="op">(</span><span class="va">tse_sub</span>, <span class="st">"pathway"</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">tse_sub</span><span class="op">)</span> <span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<p>Now, we can see that the data has been successfully subsetted, with
all sample names matching across experiments.</p>
<p>In addition to adding the entire list, we can also add individual
experiments. Let’s add the agglomerated data separately to keep the
whole dataset organized.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">altExp</span><span class="op">(</span><span class="va">tse</span>, <span class="st">"family"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">tse_family</span></span>
<span><span class="va">tse</span></span>
<span><span class="co">#&gt; class: TreeSummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 304 60 </span></span>
<span><span class="co">#&gt; metadata(1): agglomerated_by_rank</span></span>
<span><span class="co">#&gt; assays(3): counts relabundance clr</span></span>
<span><span class="co">#&gt; rownames(304): [Bacteroides] pectinophilus [Clostridium] hylemonae ...</span></span>
<span><span class="co">#&gt;   Victivallis vadensis Weissella confusa</span></span>
<span><span class="co">#&gt; rowData names(7): superkingdom phylum ... genus species</span></span>
<span><span class="co">#&gt; colnames(60): GupDM_A_11 GupDM_A_15 ... GupDM_JO GupDM_JP</span></span>
<span><span class="co">#&gt; colData names(29): study_name subject_id ... observed shannon</span></span>
<span><span class="co">#&gt; reducedDimNames(1): MDS_bray</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(3): pathway taxonomy family</span></span>
<span><span class="co">#&gt; rowLinks: a LinkDataFrame (304 rows)</span></span>
<span><span class="co">#&gt; rowTree: 1 phylo tree(s) (10430 leaves)</span></span>
<span><span class="co">#&gt; colLinks: NULL</span></span>
<span><span class="co">#&gt; colTree: NULL</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="cross-association-analysis">Cross-association analysis<a class="anchor" aria-label="anchor" href="#cross-association-analysis"></a>
</h3>
<p>Now that the entire dataset is well-organized and in one place, let’s
dive into analyzing the interconnections between experiments.
Specifically, we’ll examine whether the abundance of certain bacteria is
associated with specific metabolic activity pathway.</p>
<p>We’ll perform a simple cross-association analysis to explore these
relationships. First, we’ll preprocess the functional annotation data by
agglomerating the pathways based on each pathway’s function, and then
further by prevalence.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Agglomerate based on pathways function</span></span>
<span><span class="fu">altExp</span><span class="op">(</span><span class="va">tse</span>, <span class="st">"pathway_func"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">agglomerateByVariable</span><span class="op">(</span></span>
<span>    <span class="fu">altExp</span><span class="op">(</span><span class="va">tse</span>, <span class="st">"pathway"</span><span class="op">)</span>,</span>
<span>    by <span class="op">=</span> <span class="st">"rows"</span>,</span>
<span>    f <span class="op">=</span> <span class="st">"pathway"</span></span>
<span>    <span class="op">)</span></span>
<span><span class="co"># Agglomerate based on prevalence</span></span>
<span><span class="fu">altExp</span><span class="op">(</span><span class="va">tse</span>, <span class="st">"pathway_prev"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">agglomerateByPrevalence</span><span class="op">(</span></span>
<span>    <span class="fu">altExp</span><span class="op">(</span><span class="va">tse</span>, <span class="st">"pathway_func"</span><span class="op">)</span>,</span>
<span>    assay.type <span class="op">=</span> <span class="st">"relabundance"</span>,</span>
<span>    detection <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>    prevalence <span class="op">=</span> <span class="fl">0.2</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p>The data now includes the most prevalent pathways, each representing
a single metabolic activity. Next, we apply a transformation as the
final preprocessing step for our functional data.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">altExp</span><span class="op">(</span><span class="va">tse</span>, <span class="st">"pathway_prev"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">transformAssay</span><span class="op">(</span></span>
<span>    <span class="fu">altExp</span><span class="op">(</span><span class="va">tse</span>, <span class="st">"pathway_prev"</span><span class="op">)</span>,</span>
<span>    assay.type <span class="op">=</span> <span class="st">"relabundance"</span>,</span>
<span>    method <span class="op">=</span> <span class="st">"clr"</span></span>
<span>    <span class="op">)</span></span>
<span><span class="fu">altExp</span><span class="op">(</span><span class="va">tse</span>, <span class="st">"pathway_prev"</span><span class="op">)</span></span>
<span><span class="co">#&gt; class: SummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 24 60 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(2): relabundance clr</span></span>
<span><span class="co">#&gt; rownames(24): BRANCHED-CHAIN-AA-SYN-PWY: superpathway of branched amino</span></span>
<span><span class="co">#&gt;   acid biosynthesis DTDPRHAMSYN-PWY: dTDP-L-rhamnose biosynthesis I ...</span></span>
<span><span class="co">#&gt;   VALSYN-PWY: L-valine biosynthesis Other</span></span>
<span><span class="co">#&gt; rowData names(2): full_name pathway</span></span>
<span><span class="co">#&gt; colnames(60): GupDM_A_11 GupDM_A_15 ... GupDM_JO GupDM_JP</span></span>
<span><span class="co">#&gt; colData names(27): study_name subject_id ... disease_stage</span></span>
<span><span class="co">#&gt;   disease_location</span></span></code></pre></div>
<p>We’re now ready to perform the cross-association analysis. We will
examine the associations between microbial families and the most
abundant pathways. To visualize the results, we’ll use a heatmap.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform cross-associaton</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">getCrossAssociation</span><span class="op">(</span></span>
<span>    <span class="va">tse</span>,</span>
<span>    altexp1 <span class="op">=</span> <span class="st">"family"</span>,</span>
<span>    altexp2 <span class="op">=</span> <span class="st">"pathway_prev"</span>,</span>
<span>    assay.type1 <span class="op">=</span> <span class="st">"clr"</span>,</span>
<span>    assay.type2 <span class="op">=</span> <span class="st">"clr"</span>,</span>
<span>    mode <span class="op">=</span> <span class="st">"matrix"</span>,</span>
<span>    test.signif <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>    <span class="op">)</span></span>
<span><span class="co">#&gt; Calculating correlations...</span></span>
<span><span class="co">#&gt; altexp1: family, altexp2: pathway_prev, assay.type1: clr, col.var1: -, assay.type2: clr, col.var2: -</span></span>
<span><span class="co">#&gt; by: 1, function: stats::cor.test, method: kendall, test.signif: TRUE, p.adj.method: fdr, paired: FALSE, show.warnings: TRUE</span></span>
<span><span class="co">#&gt; Converting table into matrices...</span></span>
<span></span>
<span><span class="co"># Function for marking significant correlations with "X"</span></span>
<span><span class="va">add_signif</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">j</span>, <span class="va">i</span>, <span class="va">x</span>, <span class="va">y</span>, <span class="va">width</span>, <span class="va">height</span>, <span class="va">fill</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># If the p-value is under threshold</span></span>
<span>    <span class="kw">if</span><span class="op">(</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">p_adj</span><span class="op">[</span><span class="va">i</span>, <span class="va">j</span><span class="op">]</span><span class="op">)</span> <span class="op">&amp;</span> <span class="va">res</span><span class="op">$</span><span class="va">p_adj</span><span class="op">[</span><span class="va">i</span>, <span class="va">j</span><span class="op">]</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">)</span><span class="op">{</span></span>
<span>        <span class="co"># Print "X"</span></span>
<span>        <span class="fu">grid.shadowtext</span><span class="op">(</span></span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%s"</span>, <span class="st">"X"</span><span class="op">)</span>, <span class="va">x</span>, <span class="va">y</span>, gp <span class="op">=</span> <span class="fu">gpar</span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">8</span>, col <span class="op">=</span> <span class="st">"#f5f5f5"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Create a heatmap</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">Heatmap</span><span class="op">(</span></span>
<span>    <span class="co"># Correlation values</span></span>
<span>    <span class="va">res</span><span class="op">$</span><span class="va">cor</span>,</span>
<span>    <span class="co"># Print values to cells</span></span>
<span>    cell_fun <span class="op">=</span> <span class="va">add_signif</span>,</span>
<span>    <span class="co"># Adjust legend</span></span>
<span>    heatmap_legend_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        title <span class="op">=</span> <span class="st">"correlation"</span>, legend_height <span class="op">=</span> <span class="fu">unit</span><span class="op">(</span><span class="fl">5</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="co"># Adjust column names</span></span>
<span>    column_names_rot <span class="op">=</span> <span class="fl">45</span>,</span>
<span>    column_names_max_height <span class="op">=</span> <span class="fu">unit</span><span class="op">(</span><span class="fl">20</span>, <span class="st">"cm"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="eccb_workflow_files/figure-html/cross_association-1.png" width="1440"></p>
<p>For more information on data integration, see the following
sections:</p>
<ul>
<li><a href="https://microbiome.github.io/OMA/docs/devel/pages/cross_correlation.html" class="external-link">Cross-association</a></li>
<li><a href="https://microbiome.github.io/OMA/docs/devel/pages/multiassay_ordination.html" class="external-link">Ordination-based
Multiassay Analysis</a></li>
<li><a href="https://microbiome.github.io/OMA/docs/devel/pages/integrated_learner.html" class="external-link">Multi-assay
classification</a></li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="microbial-network-analysis">Microbial network analysis<a class="anchor" aria-label="anchor" href="#microbial-network-analysis"></a>
</h2>
<p>After having analyzed associations between microbial families and
pathways in the last section, we are now interested in the
<strong>microbial associations</strong> itself, which helps us to
understand the complex interplay of microbial communities.</p>
<p>In a typical microbiome network, <strong>nodes are ASVs or
taxa</strong> on a certain rank (here family). The nodes are connected
by <strong>edges, which represent microbial associations</strong>.
Higher edge thickness corresponds to higher edge weight. Edges with a
positive underlying association are blue, those with a negative
association are red.</p>
<div class="figure">
<img src="figures/Exemplary_microbiome_network.png" alt="An exemplary microbiome network." width="90%"><p class="caption">
An exemplary microbiome network.
</p>
</div>
<div class="section level3">
<h3 id="from-read-counts-to-networks">From read counts to networks<a class="anchor" aria-label="anchor" href="#from-read-counts-to-networks"></a>
</h3>
<p>An essential step in microbial network learning is to obtain an
adjacency matrix representing the network from a given read count
matrix:</p>
<p><img src="figures/From_counts_to_networks.png" width="90%"></p>
<p>Association estimation methods need to take the typical
characteristics of microbiome data into account. Here is a brief
<strong>overview of the typical data characteristics</strong> and common
approaches to solve them:</p>
<p><strong>Compositionality</strong></p>
<ul>
<li>Data transformation (e.g., clr or rclr)</li>
<li>See the <a href="https://microbiome.github.io/OMA/docs/devel/pages/transformation.html" class="external-link">OMA
chapter</a> on data transformations</li>
</ul>
<p><strong>Zero-inflation</strong></p>
<ul>
<li>Adding small pseudo counts (to the zeros only!)</li>
<li>More sophisticated zero imputation methods, e.g., the R package <a href="https://cran.r-project.org/web/packages/zCompositions/zCompositions.pdf" class="external-link">zCompositions</a>
</li>
</ul>
<p><strong>High dimensionality (p&gt;&gt;n)</strong></p>
<p>Sparse association estimation (sparsity assumption is needed due to
the underdetermined mathematical system). Depending on the desired type
of association, there’s a wide range of existing methods and R
packages:</p>
<ul>
<li>
<strong>Correlation:</strong>
<ul>
<li>
<a href="https://rdrr.io/github/zdk123/SpiecEasi/man/sparcc.html" class="external-link">SparCC</a>
(“Sparse Correlations for Compositional data”)</li>
<li>
<a href="https://github.com/huayingfang/CCLasso" class="external-link">CCLasso</a>
(“Correlation Inference for Compositional Data Through Lasso”)</li>
<li>
<a href="https://bioconductor.org/packages/release/bioc/html/ccrepe.html" class="external-link">CCREPE</a>
(“Compositionality Corrected by REnormalization and PErmutation”)</li>
</ul>
</li>
<li>
<strong>Conditional dependence / partial correlation:</strong>
<ul>
<li>
<a href="https://github.com/zdk123/SpiecEasi" class="external-link">SPIEC-EASI</a>
(“SParse InversE Covariance Estimation for Ecological ASsociation
Inference”)</li>
<li>
<a href="https://github.com/GraceYoon/SPRING" class="external-link">SPRING</a>
(“Semi-Parametric Rank-based approach for INference in Graphical
model”)</li>
<li><a href="https://github.com/huayingfang/gCoda" class="external-link">gCoda</a></li>
</ul>
</li>
<li>
<strong>Proportionality</strong>
<ul>
<li><a href="https://cran.r-project.org/web/packages/propr/index.html" class="external-link">propr</a></li>
</ul>
</li>
</ul>
<p>These methods are also available through the R package <a href="https://github.com/stefpeschel/NetCoMi" class="external-link">NetCoMi</a>, which we will
use to construct and analyze an association network for the gut
microbiome.</p>
<p>The complete workflow implemented in <code>NetCoMi</code> is shown in
the <a href="https://microbiome.github.io/OMA/docs/devel/pages/network_learning.html#sec-network-learning-workflow" class="external-link">OMA
chapter</a> on network learning.</p>
</div>
<div class="section level3">
<h3 id="network-learning">Network learning<a class="anchor" aria-label="anchor" href="#network-learning"></a>
</h3>
<p>The function for network construction <code>netConstruct()</code>
offers the possibility to <strong>filter</strong> samples and taxa
according to certain criteria. Here, we keep only <strong>taxa occurring
in at least 10% of the samples</strong>.</p>
<p>We use the <strong>SPIEC-EASI</strong> method to <strong>estimate
microbial associations</strong>. The approach includes two graphical
model inference procedures:</p>
<ul>
<li>Meinhausen and Bühlmann’s neighborhood selection (the MB
method)</li>
<li>Inverse covariance selection (the glasso method).</li>
</ul>
<p>We will use <strong>glasso</strong> here.</p>
<p>SPIEC-EASI uses the <strong>StARS</strong> (“Stability Approach to
Regularization Selection”) method to obtain a <strong>sparse association
matrix</strong>. We set the StARS threshold to 0.1 to get a denser graph
(the default is 0.05).</p>
<p>Since zero replacement, normalization, and sparsification (model
selection) is included in SPIEC-EASI, we set the corresponding
<code>netConstruct</code> arguments to “none”.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">net</span> <span class="op">&lt;-</span> <span class="fu">netConstruct</span><span class="op">(</span></span>
<span>    <span class="va">tse_family</span>,</span>
<span>    taxRank <span class="op">=</span> <span class="st">"family"</span>,</span>
<span>    filtTax <span class="op">=</span> <span class="st">"numbSamp"</span>,</span>
<span>    filtTaxPar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>numbSamp <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span>,</span>
<span>    measure <span class="op">=</span> <span class="st">"spieceasi"</span>,</span>
<span>    measurePar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"glasso"</span>, </span>
<span>                      <span class="co"># nlambda and rep.num should be higher (30 to 50)</span></span>
<span>                      nlambda<span class="op">=</span><span class="fl">20</span>, </span>
<span>                      pulsar.params<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>rep.num<span class="op">=</span><span class="fl">10</span>,</span>
<span>                                         thresh <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span>, <span class="co"># defines sparsity</span></span>
<span>                      norm.params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"mclr"</span>, </span>
<span>                                         types <span class="op">=</span> <span class="st">"count"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    zeroMethod <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>    normMethod <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>    sparsMethod <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>    dissFunc <span class="op">=</span> <span class="st">"signed"</span>,</span>
<span>    seed <span class="op">=</span> <span class="fl">16092024</span><span class="op">)</span></span></code></pre></div>
<p><code>netConstruct()</code> returns an object of the class
<code>microNet</code>, which contains all matrices generated during
network construction.</p>
<p>The object also contains an <strong>edge list</strong>, giving each
edge’s estimated association, dissimilarity, and adjacency. Let’s take a
quick look at the edges with the highest and lowest edge weights:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">edgelist</span> <span class="op">&lt;-</span> <span class="va">net</span><span class="op">$</span><span class="va">edgelist1</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">net</span><span class="op">$</span><span class="va">edgelist1</span><span class="op">$</span><span class="va">adja</span>, </span>
<span>                                decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">edgelist</span><span class="op">)</span></span>
<span><span class="co">#&gt;                  v1               v2      asso      diss      adja</span></span>
<span><span class="co">#&gt; 45 Odoribacteraceae    Rikenellaceae 0.3771518 0.5580538 0.4419462</span></span>
<span><span class="co">#&gt; 10  Akkermansiaceae Odoribacteraceae 0.3090037 0.5877909 0.4122091</span></span>
<span><span class="co">#&gt; 53    Rikenellaceae   Tannerellaceae 0.2994215 0.5918524 0.4081476</span></span>
<span><span class="co">#&gt; 46 Odoribacteraceae   Tannerellaceae 0.2967157 0.5929942 0.4070058</span></span>
<span><span class="co">#&gt; 17   Bacteroidaceae   Tannerellaceae 0.2913105 0.5952686 0.4047314</span></span>
<span><span class="co">#&gt; 16   Bacteroidaceae    Rikenellaceae 0.2838756 0.5983830 0.4016170</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">edgelist</span><span class="op">)</span></span>
<span><span class="co">#&gt;                  v1                  v2       asso      diss      adja</span></span>
<span><span class="co">#&gt; 54 Selenomonadaceae      Tannerellaceae -0.1088053 0.7445822 0.2554178</span></span>
<span><span class="co">#&gt; 23   Clostridiaceae    Lactobacillaceae -0.1088343 0.7445920 0.2554080</span></span>
<span><span class="co">#&gt; 40   Eubacteriaceae    Lactobacillaceae -0.1116851 0.7455485 0.2544515</span></span>
<span><span class="co">#&gt; 27   Clostridiaceae    Selenomonadaceae -0.1404570 0.7551348 0.2448652</span></span>
<span><span class="co">#&gt; 12  Akkermansiaceae    Selenomonadaceae -0.1423893 0.7557742 0.2442258</span></span>
<span><span class="co">#&gt; 41 Lactobacillaceae Methanobacteriaceae -0.2651769 0.7953543 0.2046457</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="network-analysis">Network analysis<a class="anchor" aria-label="anchor" href="#network-analysis"></a>
</h3>
<p>The <code>microNet</code> object is now passed to
<code>netAnalyze()</code> to perform <strong>network
analysis</strong>.</p>
<p>The function computes several common network characteristics such as
centrality measures, hubs, cluster assignment, and global network
measures.</p>
<p>The user has several options to choose from, such as a clustering
method, how to define hubs, and whether or not to normalize centrality
values. See the help page <code>?netAnalyze</code> for a description of
the arguments.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">netprops</span> <span class="op">&lt;-</span> <span class="fu">netAnalyze</span><span class="op">(</span><span class="va">net</span>,</span>
<span>                       clustMethod <span class="op">=</span> <span class="st">"cluster_fast_greedy"</span>,</span>
<span>                       hubPar <span class="op">=</span> <span class="st">"eigenvector"</span>,</span>
<span>                       normDeg <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                       gcmHeat <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>We use <code>NetCoMi's</code> plot function to <strong>visualize the
network</strong>.</p>
<p>In the first plot, <strong>node colors</strong> represent the
detected <strong>clusters</strong> and <strong>node sizes</strong> are
scaled by <strong>eigenvector centrality</strong>. <strong>Hub
nodes</strong> are highlighted by default. Unconnected nodes are not
included in the plot.</p>
<p>Note that nodes are sometimes placed too close together so that the
labels overlap. You may need to play around with the
<code>repulsion</code> argument until you find a value where the labels
are legible, but also the clusters are still well recognizable.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Palette with distinguishable colors</span></span>
<span><span class="va">nodecols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#E69F00"</span>, <span class="st">"#56B4E9"</span>, <span class="st">"#009E73"</span>,  <span class="st">"#CC79A7"</span>, <span class="st">"#F0E442"</span>, </span>
<span>              <span class="st">"#0072B2"</span>, <span class="st">"#D55E00"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">netprops</span>,</span>
<span>     repulsion <span class="op">=</span> <span class="fl">1</span>,</span>
<span>     rmSingles <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     shortenLabels <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>     labelScale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>     nodeSize <span class="op">=</span> <span class="st">"eigenvector"</span>,</span>
<span>     nodeSizeSpread <span class="op">=</span> <span class="fl">3</span>,</span>
<span>     nodeColor <span class="op">=</span> <span class="st">"cluster"</span>,</span>
<span>     colorVec <span class="op">=</span> <span class="va">nodecols</span>,</span>
<span>     nodeTransp <span class="op">=</span> <span class="fl">20</span>,</span>
<span>     posCol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#5C7aee"</span>, <span class="st">"#273cc2"</span><span class="op">)</span>,</span>
<span>     hubBorderCol <span class="op">=</span> <span class="st">"gray40"</span>,</span>
<span>     cexNodes <span class="op">=</span> <span class="fl">1.8</span>,</span>
<span>     edgeTranspHigh <span class="op">=</span> <span class="fl">20</span>,</span>
<span>     title1 <span class="op">=</span> <span class="st">"Network properties highlighted"</span>,</span>
<span>     showTitle <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     cexTitle <span class="op">=</span> <span class="fl">2</span>,</span>
<span>     mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">8</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="fl">0.9</span>, <span class="fl">1.1</span>, cex <span class="op">=</span> <span class="fl">1.7</span>, title <span class="op">=</span> <span class="st">"estimated correlation:"</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span>,<span class="st">"-"</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#273cc2"</span>,<span class="st">"red"</span><span class="op">)</span>,</span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>xpd <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="fl">0.7</span>, <span class="op">-</span><span class="fl">0.7</span>, cex <span class="op">=</span> <span class="fl">1.5</span>, bty <span class="op">=</span> <span class="st">"n"</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Node color: Clusters"</span>, </span>
<span>                  <span class="st">"Node size: Eigenvector centrality"</span>,</span>
<span>                  <span class="st">"Hubs: Highest eigen. centrality"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="eccb_workflow_files/figure-html/network_plot_clusters-1.png" width="100%"></p>
<p>We now <strong>color</strong> nodes according to their
<strong>phylum</strong>. The <strong>node sizes</strong> are
proportional to a taxon’s sum of <strong>mclr-transformed
abundances</strong>.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate vector with phylum names for node coloring</span></span>
<span><span class="va">phyla</span> <span class="op">&lt;-</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">tse_family</span><span class="op">)</span><span class="op">$</span><span class="va">phylum</span></span>
<span></span>
<span><span class="co"># Replace phyla that occur only once with "other"</span></span>
<span><span class="va">phylum_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">phyla</span><span class="op">)</span></span>
<span><span class="va">phyla</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">phyla</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">phylum_counts</span><span class="op">[</span><span class="va">phylum_counts</span> <span class="op">==</span> <span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">|</span> </span>
<span>                  <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">phyla</span><span class="op">)</span>, <span class="st">"Other"</span>, <span class="va">phyla</span><span class="op">)</span></span>
<span></span>
<span><span class="va">phyla</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">phyla</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Make level "Other" the last one</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">phyla</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">phyla</span><span class="op">)</span>, <span class="st">"Other"</span><span class="op">)</span>, <span class="st">"Other"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Assign family names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">phyla</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">tse_family</span><span class="op">)</span><span class="op">$</span><span class="va">family</span></span>
<span></span>
<span><span class="co"># Node colors (phylum "Other" should be gray)</span></span>
<span><span class="va">colors_phylum</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">nodecols</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nlevels.html" class="external-link">nlevels</a></span><span class="op">(</span><span class="va">phyla</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span><span class="op">]</span>, <span class="st">"#999999"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">netprops</span>,</span>
<span>     repulsion <span class="op">=</span> <span class="fl">1</span>,</span>
<span>     rmSingles <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     shortenLabels <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>     labelScale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>     nodeSize <span class="op">=</span> <span class="st">"mclr"</span>,</span>
<span>     nodeColor <span class="op">=</span> <span class="st">"feature"</span>,</span>
<span>     featVecCol <span class="op">=</span> <span class="va">phyla</span>,</span>
<span>     colorVec <span class="op">=</span> <span class="va">colors_phylum</span>,</span>
<span>     nodeTransp <span class="op">=</span> <span class="fl">20</span>,</span>
<span>     posCol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#5C7aee"</span>, <span class="st">"#273cc2"</span><span class="op">)</span>,</span>
<span>     highlightHubs <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>     cexNodes <span class="op">=</span> <span class="fl">1.8</span>,</span>
<span>     edgeWidth <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>     edgeTranspHigh <span class="op">=</span> <span class="fl">20</span>,</span>
<span>     title1 <span class="op">=</span> <span class="st">"Data features highlighted"</span>,</span>
<span>     showTitle <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     cexTitle <span class="op">=</span> <span class="fl">2</span>,</span>
<span>     mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">10</span>, <span class="fl">4</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add legends</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="fl">0.7</span>, <span class="fl">1.1</span>, cex <span class="op">=</span> <span class="fl">1.7</span>, title <span class="op">=</span> <span class="st">"estimated correlation:"</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span>,<span class="st">"-"</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">3</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#273cc2"</span>,<span class="st">"red"</span><span class="op">)</span>,</span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Colors used in the legend should be equally transparent as in the plot</span></span>
<span><span class="va">col_transp</span> <span class="op">&lt;-</span> <span class="fu">colToTransp</span><span class="op">(</span><span class="va">colors_phylum</span>, <span class="fl">20</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1.8</span>, <span class="fl">1.1</span>, cex <span class="op">=</span> <span class="fl">1.7</span>, pt.cex <span class="op">=</span> <span class="fl">2.5</span>, title <span class="op">=</span> <span class="st">"Phylum:"</span>,</span>
<span>       legend<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">phyla</span><span class="op">)</span>, col <span class="op">=</span> <span class="va">col_transp</span>, bty <span class="op">=</span> <span class="st">"n"</span>, pch <span class="op">=</span> <span class="fl">16</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>xpd <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="op">-</span><span class="fl">0.8</span>, cex <span class="op">=</span> <span class="fl">1.5</span>, bty <span class="op">=</span> <span class="st">"n"</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Node color: Phylum"</span>, </span>
<span>                  <span class="st">"Node size: Sum of mclr abundances"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="eccb_workflow_files/figure-html/network_plot_phyla-1.png" width="100%"></p>
<p>In addition to the visual analysis, we can also have a look at the
computed <strong>network properties</strong>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">netprops</span>, numbNodes <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Component sizes</span></span>
<span><span class="co">#&gt; ```````````````          </span></span>
<span><span class="co">#&gt; size: 28 1</span></span>
<span><span class="co">#&gt;    #:  1 8</span></span>
<span><span class="co">#&gt; ______________________________</span></span>
<span><span class="co">#&gt; Global network properties</span></span>
<span><span class="co">#&gt; `````````````````````````</span></span>
<span><span class="co">#&gt; Largest connected component (LCC):</span></span>
<span><span class="co">#&gt;                                  </span></span>
<span><span class="co">#&gt; Relative LCC size         0.77778</span></span>
<span><span class="co">#&gt; Clustering coefficient    0.27864</span></span>
<span><span class="co">#&gt; Modularity                0.36060</span></span>
<span><span class="co">#&gt; Positive edge percentage 77.77778</span></span>
<span><span class="co">#&gt; Edge density              0.14286</span></span>
<span><span class="co">#&gt; Natural connectivity      0.05141</span></span>
<span><span class="co">#&gt; Vertex connectivity       1.00000</span></span>
<span><span class="co">#&gt; Edge connectivity         1.00000</span></span>
<span><span class="co">#&gt; Average dissimilarity*    0.95366</span></span>
<span><span class="co">#&gt; Average path length**     1.97776</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Whole network:</span></span>
<span><span class="co">#&gt;                                  </span></span>
<span><span class="co">#&gt; Number of components      9.00000</span></span>
<span><span class="co">#&gt; Clustering coefficient    0.27864</span></span>
<span><span class="co">#&gt; Modularity                0.36060</span></span>
<span><span class="co">#&gt; Positive edge percentage 77.77778</span></span>
<span><span class="co">#&gt; Edge density              0.08571</span></span>
<span><span class="co">#&gt; Natural connectivity      0.03746</span></span>
<span><span class="co">#&gt; -----</span></span>
<span><span class="co">#&gt; *: Dissimilarity = 1 - edge weight</span></span>
<span><span class="co">#&gt; **: Path length = Units with average dissimilarity</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ______________________________</span></span>
<span><span class="co">#&gt; Clusters</span></span>
<span><span class="co">#&gt; - In the whole network</span></span>
<span><span class="co">#&gt; - Algorithm: cluster_fast_greedy</span></span>
<span><span class="co">#&gt; ```````````````````````````````` </span></span>
<span><span class="co">#&gt;                </span></span>
<span><span class="co">#&gt; name: 0 1 2 3 4</span></span>
<span><span class="co">#&gt;    #: 8 9 7 8 4</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ______________________________</span></span>
<span><span class="co">#&gt; Hubs</span></span>
<span><span class="co">#&gt; - In alphabetical/numerical order</span></span>
<span><span class="co">#&gt; - Based on empirical quantiles of centralities</span></span>
<span><span class="co">#&gt; ```````````````````````````````````````````````                 </span></span>
<span><span class="co">#&gt;  Akkermansiaceae </span></span>
<span><span class="co">#&gt;  Odoribacteraceae</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ______________________________</span></span>
<span><span class="co">#&gt; Centrality measures</span></span>
<span><span class="co">#&gt; - In decreasing order</span></span>
<span><span class="co">#&gt; - Centrality of disconnected components is zero</span></span>
<span><span class="co">#&gt; ````````````````````````````````````````````````</span></span>
<span><span class="co">#&gt; Degree (unnormalized):</span></span>
<span><span class="co">#&gt;                   </span></span>
<span><span class="co">#&gt; Akkermansiaceae  9</span></span>
<span><span class="co">#&gt; Odoribacteraceae 9</span></span>
<span><span class="co">#&gt; Clostridiaceae   8</span></span>
<span><span class="co">#&gt; Lactobacillaceae 8</span></span>
<span><span class="co">#&gt; Rikenellaceae    8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Betweenness centrality (normalized):</span></span>
<span><span class="co">#&gt;                         </span></span>
<span><span class="co">#&gt; Lactobacillaceae 0.27066</span></span>
<span><span class="co">#&gt; Tannerellaceae   0.18234</span></span>
<span><span class="co">#&gt; Odoribacteraceae 0.15954</span></span>
<span><span class="co">#&gt; Akkermansiaceae  0.14245</span></span>
<span><span class="co">#&gt; Eubacteriaceae   0.13960</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Closeness centrality (normalized):</span></span>
<span><span class="co">#&gt;                         </span></span>
<span><span class="co">#&gt; Akkermansiaceae  0.89177</span></span>
<span><span class="co">#&gt; Odoribacteraceae 0.88758</span></span>
<span><span class="co">#&gt; Rikenellaceae    0.84766</span></span>
<span><span class="co">#&gt; Clostridiaceae   0.83774</span></span>
<span><span class="co">#&gt; Lactobacillaceae 0.82564</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Eigenvector centrality (normalized):</span></span>
<span><span class="co">#&gt;                         </span></span>
<span><span class="co">#&gt; Odoribacteraceae 1.00000</span></span>
<span><span class="co">#&gt; Akkermansiaceae  0.91584</span></span>
<span><span class="co">#&gt; Rikenellaceae    0.78864</span></span>
<span><span class="co">#&gt; Tannerellaceae   0.76453</span></span>
<span><span class="co">#&gt; Clostridiaceae   0.72856</span></span></code></pre>
<p>The network consists of one large connected component (lcc) with 28
nodes and 8 unconnected nodes.</p>
<p>Akkermansiaceae and Odoribacteraceae are identified as hub nodes
(highly connected families), which are nodes with an eigenvector
centrality above the 95% quantile of all computed eigenvector centrality
values.</p>
</div>
<div class="section level3">
<h3 id="network-comparison">Network comparison<a class="anchor" aria-label="anchor" href="#network-comparison"></a>
</h3>
<p>As next step, we might be interested in whether the microbial network
differs between <strong>colorectal cancer (CRC) patients</strong> and
<strong>control subjects</strong>.</p>
<p>This task could be performed with the NetCoMi function
<code>netCompare</code>, as described in the <a href="https://microbiome.github.io/OMA/docs/devel/pages/network_comparison.html" class="external-link">OMA
chapter</a> on network comparison.</p>
</div>
</div>
<div class="section level2">
<h2 id="interactive-microbiome-data-exploration-with-iseetree">Interactive microbiome data exploration with iSEEtree<a class="anchor" aria-label="anchor" href="#interactive-microbiome-data-exploration-with-iseetree"></a>
</h2>
<p><a href="https://microbiome.github.io/iSEEtree/" class="external-link">iSEEtree</a> is a
shiny app that provides a Graphical User Interface (GUI) to visualize
and interactively explore <code>TreeSE</code> objects. We will
demonstrate its capability by reproducing some of the plots that were
previously created in this workshop, this time without the need of
coding (small lie).</p>
<p>As we did before for <code>tse</code>, we add alpha and beta
diversity metrics also to the agglomerated experiment
<code>tse_family</code>, so we will be able to visualize these metrics
in the app.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Add information on alpha diversity</span></span>
<span><span class="va">tse_family</span> <span class="op">&lt;-</span> <span class="fu">addAlpha</span><span class="op">(</span><span class="va">tse_family</span>,</span>
<span>                       index <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"coverage_diversity"</span>, <span class="st">"shannon_diversity"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add information on beta diversity</span></span>
<span><span class="va">tse_family</span> <span class="op">&lt;-</span> <span class="fu">runMDS</span><span class="op">(</span><span class="va">tse_family</span>,</span>
<span>                     FUN <span class="op">=</span> <span class="va">getDissimilarity</span>,</span>
<span>                     method <span class="op">=</span> <span class="st">"bray"</span>,</span>
<span>                     assay.type <span class="op">=</span> <span class="st">"relabundance"</span>,</span>
<span>                     ncomp <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                     name <span class="op">=</span> <span class="st">"MDS_bray"</span><span class="op">)</span></span></code></pre></div>
<p>Launching the app is easy. All you need to do is run the command
<code>iSEE</code> on the <code>TreeSE</code> object that you want to
explore.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Launch app</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/interactive.html" class="external-link">interactive</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">iSEE</span><span class="op">(</span><span class="va">tse_family</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>The default panel layout does not contain the so-called
<code>ColumnDataTable</code> panel, which plots variables from the
<code>colData.</code> Therefore, we need to specify a custom panel
layout with the <code>initial</code> argument to generate the alpha
diversity plot.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Launch app with custom layout</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/interactive.html" class="external-link">interactive</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">iSEE</span><span class="op">(</span><span class="va">tse_family</span>, initial <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu">ColumnDataTable</span><span class="op">(</span><span class="op">)</span>, <span class="fu">ColumnDataPlot</span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>If interested, you can explore further functionality and the panel
catalogue in the <a href="https://microbiome.github.io/iSEEtree/articles/iSEEtree.html" class="external-link">iSEEtree
vignettes</a>.</p>
<p>Ideas for self-learning:</p>
<ul>
<li>Learn what different panels show by reading their help page</li>
<li>Adjust a plot to deliver the message you want</li>
<li>Use selection parameters to show a subset of the data</li>
<li>Save a plot to image and/or access its R code</li>
</ul>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Etc/UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    grid      stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">#&gt; [8] methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] tidyr_1.3.1                     patchwork_1.2.0                </span></span>
<span><span class="co">#&gt;  [3] shadowtext_0.1.4                scater_1.33.4                  </span></span>
<span><span class="co">#&gt;  [5] scuttle_1.15.4                  NetCoMi_1.1.0                  </span></span>
<span><span class="co">#&gt;  [7] SpiecEasi_1.1.3                 miaViz_1.13.10                 </span></span>
<span><span class="co">#&gt;  [9] ggraph_2.2.1                    ggplot2_3.5.1                  </span></span>
<span><span class="co">#&gt; [11] mia_1.13.38                     MultiAssayExperiment_1.31.5    </span></span>
<span><span class="co">#&gt; [13] iSEEtree_0.99.6                 iSEE_2.17.4                    </span></span>
<span><span class="co">#&gt; [15] ggsignif_0.6.4                  dplyr_1.1.4                    </span></span>
<span><span class="co">#&gt; [17] curatedMetagenomicData_3.13.0   TreeSummarizedExperiment_2.13.0</span></span>
<span><span class="co">#&gt; [19] Biostrings_2.73.1               XVector_0.45.0                 </span></span>
<span><span class="co">#&gt; [21] SingleCellExperiment_1.27.2     SummarizedExperiment_1.35.1    </span></span>
<span><span class="co">#&gt; [23] Biobase_2.65.1                  GenomicRanges_1.57.1           </span></span>
<span><span class="co">#&gt; [25] GenomeInfoDb_1.41.1             IRanges_2.39.2                 </span></span>
<span><span class="co">#&gt; [27] S4Vectors_0.43.2                BiocGenerics_0.51.1            </span></span>
<span><span class="co">#&gt; [29] MatrixGenerics_1.17.0           matrixStats_1.4.1              </span></span>
<span><span class="co">#&gt; [31] ComplexHeatmap_2.21.0           BiocManager_1.30.25            </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] nnet_7.3-19                 DT_0.33                    </span></span>
<span><span class="co">#&gt;   [3] vctrs_0.6.5                 digest_0.6.37              </span></span>
<span><span class="co">#&gt;   [5] png_0.1-8                   corpcor_1.6.10             </span></span>
<span><span class="co">#&gt;   [7] shape_1.4.6.1               pcaPP_2.0-5                </span></span>
<span><span class="co">#&gt;   [9] rbiom_1.0.3                 ggrepel_0.9.6              </span></span>
<span><span class="co">#&gt;  [11] corrplot_0.94               permute_0.9-7              </span></span>
<span><span class="co">#&gt;  [13] mediation_4.5.0             SPRING_1.0.4               </span></span>
<span><span class="co">#&gt;  [15] MASS_7.3-61                 pkgdown_2.1.0              </span></span>
<span><span class="co">#&gt;  [17] reshape2_1.4.4              httpuv_1.6.15              </span></span>
<span><span class="co">#&gt;  [19] foreach_1.5.2               withr_3.0.1                </span></span>
<span><span class="co">#&gt;  [21] psych_2.4.6.26              xfun_0.47                  </span></span>
<span><span class="co">#&gt;  [23] ggfun_0.1.6                 survival_3.7-0             </span></span>
<span><span class="co">#&gt;  [25] memoise_2.0.1               ggbeeswarm_0.7.2           </span></span>
<span><span class="co">#&gt;  [27] systemfonts_1.1.0           gtools_3.9.5               </span></span>
<span><span class="co">#&gt;  [29] ragg_1.3.3                  tidytree_0.4.6             </span></span>
<span><span class="co">#&gt;  [31] zoo_1.8-12                  GlobalOptions_0.1.2        </span></span>
<span><span class="co">#&gt;  [33] pbapply_1.7-2               Formula_1.2-5              </span></span>
<span><span class="co">#&gt;  [35] KEGGREST_1.45.1             promises_1.3.0             </span></span>
<span><span class="co">#&gt;  [37] httr_1.4.7                  rhdf5filters_1.17.0        </span></span>
<span><span class="co">#&gt;  [39] rhdf5_2.49.0                rstudioapi_0.16.0          </span></span>
<span><span class="co">#&gt;  [41] shinyAce_0.4.2              UCSC.utils_1.1.0           </span></span>
<span><span class="co">#&gt;  [43] miniUI_0.1.1.1              generics_0.1.3             </span></span>
<span><span class="co">#&gt;  [45] base64enc_0.1-3             curl_5.2.2                 </span></span>
<span><span class="co">#&gt;  [47] zlibbioc_1.51.1             ScaledMatrix_1.13.0        </span></span>
<span><span class="co">#&gt;  [49] polyclip_1.10-7             doSNOW_1.0.20              </span></span>
<span><span class="co">#&gt;  [51] quadprog_1.5-8              GenomeInfoDbData_1.2.12    </span></span>
<span><span class="co">#&gt;  [53] ExperimentHub_2.13.1        SparseArray_1.5.34         </span></span>
<span><span class="co">#&gt;  [55] xtable_1.8-4                stringr_1.5.1              </span></span>
<span><span class="co">#&gt;  [57] desc_1.4.3                  ade4_1.7-22                </span></span>
<span><span class="co">#&gt;  [59] doParallel_1.0.17           evaluate_0.24.0            </span></span>
<span><span class="co">#&gt;  [61] S4Arrays_1.5.7              BiocFileCache_2.13.0       </span></span>
<span><span class="co">#&gt;  [63] preprocessCore_1.67.0       glmnet_4.1-8               </span></span>
<span><span class="co">#&gt;  [65] pulsar_0.3.11               irlba_2.3.5.1              </span></span>
<span><span class="co">#&gt;  [67] colorspace_2.1-1            filelock_1.0.3             </span></span>
<span><span class="co">#&gt;  [69] shinyWidgets_0.8.6          magrittr_2.0.3             </span></span>
<span><span class="co">#&gt;  [71] later_1.3.2                 viridis_0.6.5              </span></span>
<span><span class="co">#&gt;  [73] ggtree_3.13.1               lattice_0.22-6             </span></span>
<span><span class="co">#&gt;  [75] DECIPHER_3.1.4              cowplot_1.1.3              </span></span>
<span><span class="co">#&gt;  [77] Hmisc_5.1-3                 pillar_1.9.0               </span></span>
<span><span class="co">#&gt;  [79] nlme_3.1-166                huge_1.3.5                 </span></span>
<span><span class="co">#&gt;  [81] iterators_1.0.14            decontam_1.25.0            </span></span>
<span><span class="co">#&gt;  [83] compiler_4.4.1              beachmat_2.21.6            </span></span>
<span><span class="co">#&gt;  [85] stringi_1.8.4               biomformat_1.33.0          </span></span>
<span><span class="co">#&gt;  [87] tokenizers_0.3.0            minqa_1.2.8                </span></span>
<span><span class="co">#&gt;  [89] plyr_1.8.9                  crayon_1.5.3               </span></span>
<span><span class="co">#&gt;  [91] abind_1.4-8                 gridGraphics_0.5-1         </span></span>
<span><span class="co">#&gt;  [93] graphlayouts_1.1.1          bit_4.0.5                  </span></span>
<span><span class="co">#&gt;  [95] rootSolve_1.8.2.4           mixedCCA_1.6.2             </span></span>
<span><span class="co">#&gt;  [97] sandwich_3.1-1              fastcluster_1.2.6          </span></span>
<span><span class="co">#&gt;  [99] codetools_0.2-20            textshaping_0.4.0          </span></span>
<span><span class="co">#&gt; [101] BiocSingular_1.21.3         bslib_0.8.0                </span></span>
<span><span class="co">#&gt; [103] slam_0.1-53                 phyloseq_1.49.0            </span></span>
<span><span class="co">#&gt; [105] GetoptLong_1.0.5            tidytext_0.4.2             </span></span>
<span><span class="co">#&gt; [107] multtest_2.61.0             mime_0.12                  </span></span>
<span><span class="co">#&gt; [109] splines_4.4.1               circlize_0.4.16            </span></span>
<span><span class="co">#&gt; [111] Rcpp_1.0.13                 dbplyr_2.5.0               </span></span>
<span><span class="co">#&gt; [113] sparseMatrixStats_1.17.2    qgraph_1.9.8               </span></span>
<span><span class="co">#&gt; [115] knitr_1.48                  blob_1.2.4                 </span></span>
<span><span class="co">#&gt; [117] utf8_1.2.4                  here_1.0.1                 </span></span>
<span><span class="co">#&gt; [119] clue_0.3-65                 BiocVersion_3.20.0         </span></span>
<span><span class="co">#&gt; [121] lme4_1.1-35.5               pbivnorm_0.6.0             </span></span>
<span><span class="co">#&gt; [123] fs_1.6.4                    checkmate_2.3.2            </span></span>
<span><span class="co">#&gt; [125] DelayedMatrixStats_1.27.3   Rdpack_2.6.1               </span></span>
<span><span class="co">#&gt; [127] ggplotify_0.1.2             lavaan_0.6-18              </span></span>
<span><span class="co">#&gt; [129] tibble_3.2.1                Matrix_1.7-0               </span></span>
<span><span class="co">#&gt; [131] statmod_1.5.0               lpSolve_5.6.21             </span></span>
<span><span class="co">#&gt; [133] tweenr_2.0.3                pkgconfig_2.0.3            </span></span>
<span><span class="co">#&gt; [135] tools_4.4.1                 cachem_1.1.0               </span></span>
<span><span class="co">#&gt; [137] rbibutils_2.2.16            RSQLite_2.3.7              </span></span>
<span><span class="co">#&gt; [139] viridisLite_0.4.2           DBI_1.2.3                  </span></span>
<span><span class="co">#&gt; [141] impute_1.79.0               fastmap_1.2.0              </span></span>
<span><span class="co">#&gt; [143] rmarkdown_2.28              scales_1.3.0               </span></span>
<span><span class="co">#&gt; [145] shinydashboard_0.7.2        AnnotationHub_3.13.3       </span></span>
<span><span class="co">#&gt; [147] sass_0.4.9                  rpart_4.1.23               </span></span>
<span><span class="co">#&gt; [149] snow_0.4-4                  farver_2.1.2               </span></span>
<span><span class="co">#&gt; [151] tidygraph_1.3.1             mgcv_1.9-1                 </span></span>
<span><span class="co">#&gt; [153] yaml_2.3.10                 VGAM_1.1-11                </span></span>
<span><span class="co">#&gt; [155] foreign_0.8-87              cli_3.6.3                  </span></span>
<span><span class="co">#&gt; [157] purrr_1.0.2                 lifecycle_1.0.4            </span></span>
<span><span class="co">#&gt; [159] mvtnorm_1.3-1               bluster_1.15.1             </span></span>
<span><span class="co">#&gt; [161] backports_1.5.0             rintrojs_0.3.4             </span></span>
<span><span class="co">#&gt; [163] BiocParallel_1.39.0         gtable_0.3.5               </span></span>
<span><span class="co">#&gt; [165] rjson_0.2.22                limma_3.61.9               </span></span>
<span><span class="co">#&gt; [167] parallel_4.4.1              ape_5.8                    </span></span>
<span><span class="co">#&gt; [169] SnowballC_0.7.1             jsonlite_1.8.8             </span></span>
<span><span class="co">#&gt; [171] colourpicker_1.3.0          bit64_4.0.5                </span></span>
<span><span class="co">#&gt; [173] glasso_1.11                 yulab.utils_0.1.7          </span></span>
<span><span class="co">#&gt; [175] vegan_2.6-8                 BiocNeighbors_1.99.0       </span></span>
<span><span class="co">#&gt; [177] RcppParallel_5.1.9          janeaustenr_1.0.0          </span></span>
<span><span class="co">#&gt; [179] jquerylib_0.1.4             highr_0.11                 </span></span>
<span><span class="co">#&gt; [181] shinyjs_2.1.0               orca_1.1-2                 </span></span>
<span><span class="co">#&gt; [183] lazyeval_0.2.2              shiny_1.9.1                </span></span>
<span><span class="co">#&gt; [185] dynamicTreeCut_1.63-1       htmltools_0.5.8.1          </span></span>
<span><span class="co">#&gt; [187] GO.db_3.19.1                rappdirs_0.3.3             </span></span>
<span><span class="co">#&gt; [189] glue_1.7.0                  rprojroot_2.0.4            </span></span>
<span><span class="co">#&gt; [191] treeio_1.29.1               mnormt_2.1.1               </span></span>
<span><span class="co">#&gt; [193] jpeg_0.1-10                 gridExtra_2.3              </span></span>
<span><span class="co">#&gt; [195] boot_1.3-31                 igraph_2.0.3               </span></span>
<span><span class="co">#&gt; [197] R6_2.5.1                    fdrtool_1.2.18             </span></span>
<span><span class="co">#&gt; [199] labeling_0.4.3              listviewer_4.0.0           </span></span>
<span><span class="co">#&gt; [201] cluster_2.1.6               Rhdf5lib_1.27.0            </span></span>
<span><span class="co">#&gt; [203] aplot_0.2.3                 nloptr_2.1.1               </span></span>
<span><span class="co">#&gt; [205] DirichletMultinomial_1.47.0 DelayedArray_0.31.11       </span></span>
<span><span class="co">#&gt; [207] tidyselect_1.2.1            vipor_0.4.7                </span></span>
<span><span class="co">#&gt; [209] WGCNA_1.72-5                htmlTable_2.4.3            </span></span>
<span><span class="co">#&gt; [211] ggforce_0.4.2               AnnotationDbi_1.67.0       </span></span>
<span><span class="co">#&gt; [213] filematrix_1.3              rsvd_1.0.5                 </span></span>
<span><span class="co">#&gt; [215] munsell_0.5.1               data.table_1.16.0          </span></span>
<span><span class="co">#&gt; [217] htmlwidgets_1.6.4           RColorBrewer_1.1-3         </span></span>
<span><span class="co">#&gt; [219] rlang_1.1.4                 Cairo_1.6-2                </span></span>
<span><span class="co">#&gt; [221] ggnewscale_0.5.0            fansi_1.0.6                </span></span>
<span><span class="co">#&gt; [223] beeswarm_0.4.0</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Tuomas Borman, Leo Lahti.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
